From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: zhangyuheng <zhangyuheng@lunadeer.cn>
Date: Tue, 11 Mar 2025 14:31:11 +0800
Subject: [PATCH] Dynamic Activation Brain


diff --git a/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java b/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java
index 610abaee663b7573a56c0a0a6dd4ab9c0224b173..cebf0a06f4be71515106b384e9b38c4f4a786e13 100644
--- a/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java
+++ b/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java
@@ -416,4 +416,13 @@ public class GlobalConfiguration extends ConfigurationPart {
         }
     }
     // Folia end - threaded regions
+    // DeerFolia start - puffish DAB (Dynamic Activation Brain)
+    public DynamicActivationBrain dynamicActivationBrain;
+    public class DynamicActivationBrain extends ConfigurationPart {
+        public boolean dearEnabled = true;
+        public int startDistance = 12;
+        public int maximumActivationPrio = 20;
+        public int activationDistanceMod = 8;
+    }
+    // DeerFolia end - puffish DAB
 }
