From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: zhangyuheng <zhangyuheng@lunadeer.cn>
Date: Tue, 11 Mar 2025 18:04:31 +0800
Subject: [PATCH] Async pathfinding


diff --git a/net/minecraft/world/entity/ai/behavior/AcquirePoi.java b/net/minecraft/world/entity/ai/behavior/AcquirePoi.java
index e2eb02d8a0a9481be2308c9052582591bc4f9850..1778f5b62020fb0084d47664d4b957d239a12d79 100644
--- a/net/minecraft/world/entity/ai/behavior/AcquirePoi.java
+++ b/net/minecraft/world/entity/ai/behavior/AcquirePoi.java
@@ -104,6 +104,30 @@ public class AcquirePoi {
                             }
                         }
                         // Paper end - optimise POI searches
+                        // DeerFolia start - petal - Async path processing
+                        if (cn.lunadeer.mc.deerfolia.DeerFoliaConfiguration.asyncPathfinding.enabled) {
+                            Path possiblePath = findPathToPois(body, poiPositions);
+                            cn.lunadeer.mc.deerfolia.async.path.AsyncPathProcessor.awaitProcessing(body, possiblePath, path -> {
+                                if (path == null || !path.canReach()) {
+                                    for (Pair<Holder<PoiType>, BlockPos> p : poiPositions) {
+                                        batchCache.computeIfAbsent(
+                                            p.getSecond().asLong(), key -> new AcquirePoi.JitteredLinearRetry(random, timestamp)
+                                        );
+                                    }
+                                    return;
+                                }
+
+                                BlockPos targetPos = path.getTarget();
+                                poiManager.getTypeWithoutLoad(targetPos).ifPresent(type -> {
+                                    poiManager.take(poiType, (t, poiPos) -> poiPos.equals(targetPos), targetPos, 1);
+                                    toAcquire.set(GlobalPos.of(level.dimension(), targetPos));
+                                    onPoiAcquisitionEvent.ifPresent(event -> level.broadcastEntityEvent(body, event));
+                                    batchCache.clear();
+                                    level.debugSynchronizers().updatePoi(targetPos);
+                                });
+                            });
+                        } else {
+                        // DeerFolia end
                         Path path = findPathToPois(body, poiPositions);
                         if (path != null && path.canReach()) {
                             BlockPos targetPos = path.getTarget();
@@ -119,6 +143,7 @@ public class AcquirePoi {
                                 batchCache.computeIfAbsent(p.getSecond().asLong(), key -> new AcquirePoi.JitteredLinearRetry(random, timestamp));
                             }
                         }
+                        } // DeerFolia - async path processing
 
                         return true;
                     }
diff --git a/net/minecraft/world/entity/ai/behavior/MoveToTargetSink.java b/net/minecraft/world/entity/ai/behavior/MoveToTargetSink.java
index b40da004b5281a041ee896ae176bfb9c4660f353..410cc9126098584daf821316435a88751277defd 100644
--- a/net/minecraft/world/entity/ai/behavior/MoveToTargetSink.java
+++ b/net/minecraft/world/entity/ai/behavior/MoveToTargetSink.java
@@ -21,6 +21,7 @@ public class MoveToTargetSink extends Behavior<Mob> {
     private int remainingCooldown;
     private @Nullable Path path;
     private @Nullable BlockPos lastTargetPos;
+    private boolean finishedProcessing; // DeerFolia - petal - track when path is processed
     private float speedModifier;
 
     public MoveToTargetSink() {
@@ -52,9 +53,13 @@ public class MoveToTargetSink extends Behavior<Mob> {
         Brain<?> brain = body.getBrain();
         WalkTarget walkTarget = brain.getMemory(MemoryModuleType.WALK_TARGET).get();
         boolean reachedTarget = this.reachedTarget(body, walkTarget);
-        if (!reachedTarget && this.tryComputePath(body, walkTarget, level.getGameTime())) {
+        if (!cn.lunadeer.mc.deerfolia.DeerFoliaConfiguration.asyncPathfinding.enabled
+            && !reachedTarget
+            && this.tryComputePath(body, walkTarget, level.getGameTime())) { // DeerFolia - petal - async path processing means we can't know if the path is reachable here
             this.lastTargetPos = walkTarget.getTarget().currentBlockPosition();
             return true;
+        } else if (cn.lunadeer.mc.deerfolia.DeerFoliaConfiguration.asyncPathfinding.enabled && !reachedTarget) {
+            return true; // DeerFolia - async pathfinding
         }
 
         brain.eraseMemory(MemoryModuleType.WALK_TARGET);
@@ -67,6 +72,7 @@ public class MoveToTargetSink extends Behavior<Mob> {
 
     @Override
     protected boolean canStillUse(final ServerLevel level, final Mob body, final long timestamp) {
+        if (cn.lunadeer.mc.deerfolia.DeerFoliaConfiguration.asyncPathfinding.enabled && !this.finishedProcessing) return true; // DeerFolia - petal - wait for processing
         if (this.path != null && this.lastTargetPos != null) {
             Optional<WalkTarget> walkTarget = body.getBrain().getMemory(MemoryModuleType.WALK_TARGET);
             boolean isSpectator = walkTarget.map(MoveToTargetSink::isWalkTargetSpectator).orElse(false);
@@ -93,27 +99,97 @@ public class MoveToTargetSink extends Behavior<Mob> {
 
     @Override
     protected void start(final ServerLevel level, final Mob body, final long timestamp) {
+        // DeerFolia start - petal - start processing
+        if (cn.lunadeer.mc.deerfolia.DeerFoliaConfiguration.asyncPathfinding.enabled) {
+            Brain<?> brain = body.getBrain();
+            WalkTarget walkTarget = brain.getMemory(MemoryModuleType.WALK_TARGET).get();
+
+            this.finishedProcessing = false;
+            this.lastTargetPos = walkTarget.getTarget().currentBlockPosition();
+            this.path = this.computePath(body, walkTarget);
+            return;
+        }
+        // DeerFolia end
         body.getBrain().setMemory(MemoryModuleType.PATH, this.path);
         body.getNavigation().moveTo(this.path, this.speedModifier);
     }
 
     @Override
     protected void tick(final ServerLevel level, final Mob body, final long timestamp) {
-        Path newPath = body.getNavigation().getPath();
-        Brain<?> brain = body.getBrain();
-        if (this.path != newPath) {
-            this.path = newPath;
-            brain.setMemory(MemoryModuleType.PATH, newPath);
-        }
+        // DeerFolia start - petal - Async path processing
+        if (cn.lunadeer.mc.deerfolia.DeerFoliaConfiguration.asyncPathfinding.enabled) {
+            if (this.path != null && !this.path.isProcessed()) return; // wait for processing
+
+            if (!this.finishedProcessing) {
+                this.finishedProcessing = true;
+
+                Brain<?> brain = body.getBrain();
+                boolean canReach = this.path != null && this.path.canReach();
+                if (canReach) {
+                    brain.eraseMemory(MemoryModuleType.CANT_REACH_WALK_TARGET_SINCE);
+                } else if (!brain.hasMemoryValue(MemoryModuleType.CANT_REACH_WALK_TARGET_SINCE)) {
+                    brain.setMemory(MemoryModuleType.CANT_REACH_WALK_TARGET_SINCE, timestamp);
+                }
+
+                if (!canReach) {
+                    Optional<WalkTarget> walkTarget = brain.getMemory(MemoryModuleType.WALK_TARGET);
+                    if (walkTarget.isEmpty()) {
+                        return;
+                    }
+
+                    BlockPos targetPos = walkTarget.get().getTarget().currentBlockPosition();
+                    Vec3 partialStep = DefaultRandomPos.getPosTowards((PathfinderMob)body, 10, 7, Vec3.atBottomCenterOf(targetPos), (float)(Math.PI / 2));
+                    if (partialStep != null) {
+                        this.path = body.getNavigation().createPath(partialStep.x, partialStep.y, partialStep.z, 0);
+                        this.finishedProcessing = false;
+                        return;
+                    }
+                }
+
+                body.getBrain().setMemory(MemoryModuleType.PATH, this.path);
+                body.getNavigation().moveTo(this.path, this.speedModifier);
+            }
 
-        if (newPath != null && this.lastTargetPos != null) {
-            WalkTarget walkTarget = brain.getMemory(MemoryModuleType.WALK_TARGET).get();
-            if (walkTarget.getTarget().currentBlockPosition().distSqr(this.lastTargetPos) > 4.0 && this.tryComputePath(body, walkTarget, level.getGameTime())) {
-                this.lastTargetPos = walkTarget.getTarget().currentBlockPosition();
-                this.start(level, body, timestamp);
+            Path path = body.getNavigation().getPath();
+            Brain<?> brain = body.getBrain();
+            if (path != null && this.lastTargetPos != null && brain.hasMemoryValue(MemoryModuleType.WALK_TARGET)) {
+                WalkTarget walkTarget = brain.getMemory(MemoryModuleType.WALK_TARGET).get();
+                if (walkTarget.getTarget().currentBlockPosition().distSqr(this.lastTargetPos) > 4.0) {
+                    this.start(level, body, timestamp);
+                }
+            }
+        } else {
+        // DeerFolia end
+            Path newPath = body.getNavigation().getPath();
+            Brain<?> brain = body.getBrain();
+            if (this.path != newPath) {
+                this.path = newPath;
+                brain.setMemory(MemoryModuleType.PATH, newPath);
             }
+
+            if (newPath != null && this.lastTargetPos != null) {
+                WalkTarget walkTarget = brain.getMemory(MemoryModuleType.WALK_TARGET).get();
+                if (walkTarget.getTarget().currentBlockPosition().distSqr(this.lastTargetPos) > 4.0
+                    && this.tryComputePath(body, walkTarget, level.getGameTime())) {
+                    this.lastTargetPos = walkTarget.getTarget().currentBlockPosition();
+                    this.start(level, body, timestamp);
+                }
+            }
+        } // DeerFolia - async path processing
+    }
+
+    // DeerFolia start - petal - Async path processing
+    private @Nullable Path computePath(final Mob body, final WalkTarget walkTarget) {
+        BlockPos targetPos = walkTarget.getTarget().currentBlockPosition();
+        this.speedModifier = walkTarget.getSpeedModifier();
+        Brain<?> brain = body.getBrain();
+        if (this.reachedTarget(body, walkTarget)) {
+            brain.eraseMemory(MemoryModuleType.CANT_REACH_WALK_TARGET_SINCE);
         }
+
+        return body.getNavigation().createPath(targetPos, 0);
     }
+    // DeerFolia end
 
     private boolean tryComputePath(final Mob body, final WalkTarget walkTarget, final long timestamp) {
         BlockPos targetPos = walkTarget.getTarget().currentBlockPosition();
diff --git a/net/minecraft/world/entity/ai/behavior/SetClosestHomeAsWalkTarget.java b/net/minecraft/world/entity/ai/behavior/SetClosestHomeAsWalkTarget.java
index 5bd7473b0db1a87f905c01c19ddf99e3a447bcfa..16580f5901ffb60d9bad8d1bad7a6ca691e85c1b 100644
--- a/net/minecraft/world/entity/ai/behavior/SetClosestHomeAsWalkTarget.java
+++ b/net/minecraft/world/entity/ai/behavior/SetClosestHomeAsWalkTarget.java
@@ -62,6 +62,27 @@ public class SetClosestHomeAsWalkTarget {
                                     p -> p.is(PoiTypes.HOME), cacheTest, body.blockPosition(), AcquirePoi.SCAN_RANGE, PoiManager.Occupancy.ANY
                                 )
                                 .collect(Collectors.toSet());
+                            // DeerFolia start - petal - Async path processing
+                            if (cn.lunadeer.mc.deerfolia.DeerFoliaConfiguration.asyncPathfinding.enabled) {
+                                Path possiblePath = AcquirePoi.findPathToPois(body, pois);
+                                cn.lunadeer.mc.deerfolia.async.path.AsyncPathProcessor.awaitProcessing(body, possiblePath, path -> {
+                                    if ((path == null || !path.canReach()) && triedCount.intValue() < BATCH_SIZE) {
+                                        batchCache.long2LongEntrySet().removeIf(entry -> entry.getLongValue() < lastUpdate.longValue());
+                                        return;
+                                    }
+                                    if (path == null) {
+                                        return;
+                                    }
+
+                                    BlockPos targetPos = path.getTarget();
+                                    Optional<Holder<PoiType>> type = poiManager.getTypeWithoutLoad(targetPos);
+                                    if (type.isPresent()) {
+                                        walkTarget.set(new WalkTarget(targetPos, speedModifier, 1));
+                                        level.debugSynchronizers().updatePoi(targetPos);
+                                    }
+                                });
+                            } else {
+                            // DeerFolia end
                             Path path = AcquirePoi.findPathToPois(body, pois);
                             if (path != null && path.canReach()) {
                                 BlockPos targetPos = path.getTarget();
@@ -73,6 +94,7 @@ public class SetClosestHomeAsWalkTarget {
                             } else if (triedCount.intValue() < BATCH_SIZE) {
                                 batchCache.long2LongEntrySet().removeIf(entry -> entry.getLongValue() < lastUpdate.longValue());
                             }
+                            } // DeerFolia - async path processing
 
                             return true;
                         } else {
diff --git a/net/minecraft/world/entity/ai/goal/DoorInteractGoal.java b/net/minecraft/world/entity/ai/goal/DoorInteractGoal.java
index 0b383b16145d44e5dc29b8de123546656b4793dd..f036355d2debaa01ab03154d43419b51793e7d30 100644
--- a/net/minecraft/world/entity/ai/goal/DoorInteractGoal.java
+++ b/net/minecraft/world/entity/ai/goal/DoorInteractGoal.java
@@ -57,7 +57,7 @@ public abstract class DoorInteractGoal extends Goal {
         }
 
         Path path = this.mob.getNavigation().getPath();
-        if (path != null && !path.isDone()) {
+        if (path != null && path.isProcessed() && !path.isDone()) { // DeerFolia - async pathfinding - ensure path is processed
             for (int i = 0; i < Math.min(path.getNextNodeIndex() + 2, path.getNodeCount()); i++) {
                 Node node = path.getNode(i);
                 this.doorPos = new BlockPos(node.x, node.y + 1, node.z);
diff --git a/net/minecraft/world/entity/ai/navigation/AmphibiousPathNavigation.java b/net/minecraft/world/entity/ai/navigation/AmphibiousPathNavigation.java
index 45942481a972f2308eae80f6f0f8b886e47016ec..b5c726bc060e00593d49604a49009b2c27f74d2c 100644
--- a/net/minecraft/world/entity/ai/navigation/AmphibiousPathNavigation.java
+++ b/net/minecraft/world/entity/ai/navigation/AmphibiousPathNavigation.java
@@ -12,9 +12,25 @@ public class AmphibiousPathNavigation extends PathNavigation {
         super(mob, level);
     }
 
+    // DeerFolia start - petal - async path processing
+    private static final cn.lunadeer.mc.deerfolia.async.path.NodeEvaluatorGenerator nodeEvaluatorGenerator = nodeEvaluatorFeatures -> {
+        AmphibiousNodeEvaluator nodeEvaluator = new AmphibiousNodeEvaluator(false);
+        nodeEvaluator.setCanPassDoors(nodeEvaluatorFeatures.canPassDoors());
+        nodeEvaluator.setCanFloat(nodeEvaluatorFeatures.canFloat());
+        nodeEvaluator.setCanWalkOverFences(nodeEvaluatorFeatures.canWalkOverFences());
+        nodeEvaluator.setCanOpenDoors(nodeEvaluatorFeatures.canOpenDoors());
+        return nodeEvaluator;
+    };
+    // DeerFolia end
+
     @Override
     protected PathFinder createPathFinder(final int maxVisitedNodes) {
         this.nodeEvaluator = new AmphibiousNodeEvaluator(false);
+        // DeerFolia start - petal - async path processing
+        if (cn.lunadeer.mc.deerfolia.DeerFoliaConfiguration.asyncPathfinding.enabled) {
+            return new PathFinder(this.nodeEvaluator, maxVisitedNodes, nodeEvaluatorGenerator);
+        }
+        // DeerFolia end
         return new PathFinder(this.nodeEvaluator, maxVisitedNodes);
     }
 
diff --git a/net/minecraft/world/entity/ai/navigation/FlyingPathNavigation.java b/net/minecraft/world/entity/ai/navigation/FlyingPathNavigation.java
index e44814cfb6afb594456b8215bd13a92e93de2c85..a87622eb56a224eeeec26351ca4890b000816f94 100644
--- a/net/minecraft/world/entity/ai/navigation/FlyingPathNavigation.java
+++ b/net/minecraft/world/entity/ai/navigation/FlyingPathNavigation.java
@@ -15,9 +15,25 @@ public class FlyingPathNavigation extends PathNavigation {
         super(mob, level);
     }
 
+    // DeerFolia start - petal - async path processing
+    private static final cn.lunadeer.mc.deerfolia.async.path.NodeEvaluatorGenerator nodeEvaluatorGenerator = nodeEvaluatorFeatures -> {
+        FlyNodeEvaluator nodeEvaluator = new FlyNodeEvaluator();
+        nodeEvaluator.setCanPassDoors(nodeEvaluatorFeatures.canPassDoors());
+        nodeEvaluator.setCanFloat(nodeEvaluatorFeatures.canFloat());
+        nodeEvaluator.setCanWalkOverFences(nodeEvaluatorFeatures.canWalkOverFences());
+        nodeEvaluator.setCanOpenDoors(nodeEvaluatorFeatures.canOpenDoors());
+        return nodeEvaluator;
+    };
+    // DeerFolia end
+
     @Override
     protected PathFinder createPathFinder(final int maxVisitedNodes) {
         this.nodeEvaluator = new FlyNodeEvaluator();
+        // DeerFolia start - petal - async path processing
+        if (cn.lunadeer.mc.deerfolia.DeerFoliaConfiguration.asyncPathfinding.enabled) {
+            return new PathFinder(this.nodeEvaluator, maxVisitedNodes, nodeEvaluatorGenerator);
+        }
+        // DeerFolia end
         return new PathFinder(this.nodeEvaluator, maxVisitedNodes);
     }
 
@@ -47,6 +63,7 @@ public class FlyingPathNavigation extends PathNavigation {
         if (this.hasDelayedRecomputation) {
             this.recomputePath();
         }
+        if (this.path != null && !this.path.isProcessed()) return; // DeerFolia - petal - async path processing
 
         if (!this.isDone()) {
             if (this.canUpdatePath()) {
diff --git a/net/minecraft/world/entity/ai/navigation/GroundPathNavigation.java b/net/minecraft/world/entity/ai/navigation/GroundPathNavigation.java
index 48b93678b85c6296f6ae57876926dc6763e84faa..e30c9aa7d72db630d3ac566673556aa9ac401aec 100644
--- a/net/minecraft/world/entity/ai/navigation/GroundPathNavigation.java
+++ b/net/minecraft/world/entity/ai/navigation/GroundPathNavigation.java
@@ -24,9 +24,25 @@ public class GroundPathNavigation extends PathNavigation {
         super(mob, level);
     }
 
+    // DeerFolia start - petal - async path processing
+    protected static final cn.lunadeer.mc.deerfolia.async.path.NodeEvaluatorGenerator nodeEvaluatorGenerator = nodeEvaluatorFeatures -> {
+        WalkNodeEvaluator nodeEvaluator = new WalkNodeEvaluator();
+        nodeEvaluator.setCanPassDoors(nodeEvaluatorFeatures.canPassDoors());
+        nodeEvaluator.setCanFloat(nodeEvaluatorFeatures.canFloat());
+        nodeEvaluator.setCanWalkOverFences(nodeEvaluatorFeatures.canWalkOverFences());
+        nodeEvaluator.setCanOpenDoors(nodeEvaluatorFeatures.canOpenDoors());
+        return nodeEvaluator;
+    };
+    // DeerFolia end
+
     @Override
     protected PathFinder createPathFinder(final int maxVisitedNodes) {
         this.nodeEvaluator = new WalkNodeEvaluator();
+        // DeerFolia start - petal - async path processing
+        if (cn.lunadeer.mc.deerfolia.DeerFoliaConfiguration.asyncPathfinding.enabled) {
+            return new PathFinder(this.nodeEvaluator, maxVisitedNodes, nodeEvaluatorGenerator);
+        }
+        // DeerFolia end
         return new PathFinder(this.nodeEvaluator, maxVisitedNodes);
     }
 
diff --git a/net/minecraft/world/entity/ai/navigation/PathNavigation.java b/net/minecraft/world/entity/ai/navigation/PathNavigation.java
index 2ea1ec39a37899ae1510d0036aa670e758077537..7da9bb2aa772e4b0206cf61754b133826504547d 100644
--- a/net/minecraft/world/entity/ai/navigation/PathNavigation.java
+++ b/net/minecraft/world/entity/ai/navigation/PathNavigation.java
@@ -167,6 +167,14 @@ public abstract class PathNavigation {
             return null;
         }
 
+        // DeerFolia start - petal - catch early if it's still processing these positions let it keep processing
+        if (this.path instanceof cn.lunadeer.mc.deerfolia.async.path.AsyncPath asyncPath
+            && !asyncPath.isProcessed()
+            && asyncPath.hasSameProcessingPositions(targets)) {
+            return this.path;
+        }
+        // DeerFolia end
+
         if (this.path != null && !this.path.isDone() && targets.contains(this.targetPos)) {
             return this.path;
         }
@@ -195,11 +203,29 @@ public abstract class PathNavigation {
         PathNavigationRegion region = new PathNavigationRegion(this.level, fromPos.offset(-radius, -radius, -radius), fromPos.offset(radius, radius, radius));
         Path path = this.pathFinder.findPath(region, this.mob, targets, maxPathLength, reachRange, this.maxVisitedNodesMultiplier);
         profiler.pop();
+        // DeerFolia start - petal - async path processing
+        if (cn.lunadeer.mc.deerfolia.DeerFoliaConfiguration.asyncPathfinding.enabled) {
+            if (!targets.isEmpty()) {
+                this.targetPos = targets.iterator().next();
+            }
+            cn.lunadeer.mc.deerfolia.async.path.AsyncPathProcessor.awaitProcessing(this.mob, path, processedPath -> {
+                if (processedPath != this.path) {
+                    return;
+                }
+                if (processedPath != null && processedPath.getTarget() != null) {
+                    this.targetPos = processedPath.getTarget();
+                    this.reachRange = reachRange;
+                    this.resetStuckTimeout();
+                }
+            });
+        } else {
+        // DeerFolia end
         if (path != null && path.getTarget() != null) {
             this.targetPos = path.getTarget();
             this.reachRange = reachRange;
             this.resetStuckTimeout();
         }
+        } // DeerFolia - async path processing
 
         return path;
     }
@@ -251,8 +277,8 @@ public abstract class PathNavigation {
             return false;
         }
 
-        this.trimPath();
-        if (this.path.getNodeCount() <= 0) {
+        if (this.path.isProcessed()) this.trimPath(); // DeerFolia - petal - only trim if processed
+        if (this.path.isProcessed() && this.path.getNodeCount() <= 0) { // DeerFolia - petal - only check node count if processed
             return false;
         }
 
@@ -272,6 +298,7 @@ public abstract class PathNavigation {
         if (this.hasDelayedRecomputation) {
             this.recomputePath();
         }
+        if (this.path != null && !this.path.isProcessed()) return; // DeerFolia - petal - skip pathfinding if we're still processing
 
         if (!this.isDone()) {
             if (this.canUpdatePath()) {
@@ -297,6 +324,7 @@ public abstract class PathNavigation {
     }
 
     protected void followThePath() {
+        if (!this.path.isProcessed()) return; // DeerFolia - petal - skip if not processed
         Vec3 mobPos = this.getTempMobPos();
         this.maxDistanceToWaypoint = this.mob.getBbWidth() > 0.75F ? this.mob.getBbWidth() / 2.0F : 0.75F - this.mob.getBbWidth() / 2.0F;
         Vec3i currentNodePos = this.path.getNextNodePos();
@@ -457,7 +485,7 @@ public abstract class PathNavigation {
     public boolean shouldRecomputePath(final BlockPos pos) {
         if (this.hasDelayedRecomputation) {
             return false;
-        } else if (this.path != null && !this.path.isDone() && this.path.getNodeCount() != 0) {
+        } else if (this.path != null && this.path.isProcessed() && !this.path.isDone() && this.path.getNodeCount() != 0) { // DeerFolia - petal - Skip if not processed
             Node target = this.path.getEndNode();
             Vec3 middlePos = new Vec3((target.x + this.mob.getX()) / 2.0, (target.y + this.mob.getY()) / 2.0, (target.z + this.mob.getZ()) / 2.0);
             return pos.closerToCenterThan(middlePos, this.path.getNodeCount() - this.path.getNextNodeIndex());
diff --git a/net/minecraft/world/entity/ai/navigation/WaterBoundPathNavigation.java b/net/minecraft/world/entity/ai/navigation/WaterBoundPathNavigation.java
index 87ccac321c3a087158602d87e4d05b3273545cf1..919d76ca183ee5c2fbe5e07b01eac3f260fe2400 100644
--- a/net/minecraft/world/entity/ai/navigation/WaterBoundPathNavigation.java
+++ b/net/minecraft/world/entity/ai/navigation/WaterBoundPathNavigation.java
@@ -15,11 +15,27 @@ public class WaterBoundPathNavigation extends PathNavigation {
         super(mob, level);
     }
 
+    // DeerFolia start - petal - async path processing
+    private static final cn.lunadeer.mc.deerfolia.async.path.NodeEvaluatorGenerator nodeEvaluatorGenerator = nodeEvaluatorFeatures -> {
+        SwimNodeEvaluator nodeEvaluator = new SwimNodeEvaluator(nodeEvaluatorFeatures.allowBreaching());
+        nodeEvaluator.setCanPassDoors(nodeEvaluatorFeatures.canPassDoors());
+        nodeEvaluator.setCanFloat(nodeEvaluatorFeatures.canFloat());
+        nodeEvaluator.setCanWalkOverFences(nodeEvaluatorFeatures.canWalkOverFences());
+        nodeEvaluator.setCanOpenDoors(nodeEvaluatorFeatures.canOpenDoors());
+        return nodeEvaluator;
+    };
+    // DeerFolia end
+
     @Override
     protected PathFinder createPathFinder(final int maxVisitedNodes) {
         this.allowBreaching = this.mob.is(EntityTypes.DOLPHIN);
         this.nodeEvaluator = new SwimNodeEvaluator(this.allowBreaching);
         this.nodeEvaluator.setCanPassDoors(false);
+        // DeerFolia start - async path processing
+        if (cn.lunadeer.mc.deerfolia.DeerFoliaConfiguration.asyncPathfinding.enabled) {
+            return new PathFinder(this.nodeEvaluator, maxVisitedNodes, nodeEvaluatorGenerator);
+        }
+        // DeerFolia end
         return new PathFinder(this.nodeEvaluator, maxVisitedNodes);
     }
 
diff --git a/net/minecraft/world/entity/ai/sensing/NearestBedSensor.java b/net/minecraft/world/entity/ai/sensing/NearestBedSensor.java
index 55940fbe5c6e88b27c4322cea6d971d15dfa196a..01e37dbb24c48e8dba02a4c551d8a3dda80c651b 100644
--- a/net/minecraft/world/entity/ai/sensing/NearestBedSensor.java
+++ b/net/minecraft/world/entity/ai/sensing/NearestBedSensor.java
@@ -59,6 +59,26 @@ public class NearestBedSensor extends Sensor<Mob> {
                     e -> e.is(PoiTypes.HOME), cacheTest, body.blockPosition(), AcquirePoi.SCAN_RANGE, PoiManager.Occupancy.ANY
                 )
                 .collect(Collectors.toSet());
+            // DeerFolia start - await on async path processing
+            if (cn.lunadeer.mc.deerfolia.DeerFoliaConfiguration.asyncPathfinding.enabled) {
+                Path possiblePath = AcquirePoi.findPathToPois(body, pois);
+                cn.lunadeer.mc.deerfolia.async.path.AsyncPathProcessor.awaitProcessing(body, possiblePath, path -> {
+                    if ((path == null || !path.canReach()) && this.triedCount < 5) {
+                        this.batchCache.long2LongEntrySet().removeIf(entry -> entry.getLongValue() < this.lastUpdate);
+                        return;
+                    }
+                    if (path == null) {
+                        return;
+                    }
+
+                    BlockPos targetPos = path.getTarget();
+                    Optional<Holder<PoiType>> type = poiManager.getTypeWithoutLoad(targetPos);
+                    if (type.isPresent()) {
+                        body.getBrain().setMemory(MemoryModuleType.NEAREST_BED, targetPos);
+                    }
+                });
+            } else {
+            // DeerFolia end
             Path path = AcquirePoi.findPathToPois(body, pois);
             if (path != null && path.canReach()) {
                 BlockPos targetPos = path.getTarget();
@@ -69,6 +89,7 @@ public class NearestBedSensor extends Sensor<Mob> {
             } else if (this.triedCount < 5) {
                 this.batchCache.long2LongEntrySet().removeIf(entry -> entry.getLongValue() < this.lastUpdate);
             }
+            } // DeerFolia - async path processing
         }
     }
 }
diff --git a/net/minecraft/world/entity/ai/village/poi/PoiManager.java b/net/minecraft/world/entity/ai/village/poi/PoiManager.java
index 816fae86e0ff6e7a7abe330dd8df3928f8309307..74d23b2ed7ee228cfad5b48a5880e469850d3f40 100644
--- a/net/minecraft/world/entity/ai/village/poi/PoiManager.java
+++ b/net/minecraft/world/entity/ai/village/poi/PoiManager.java
@@ -45,6 +45,11 @@ public class PoiManager extends SectionStorage<PoiSection, PoiSection.Packed> im
     public static final int VILLAGE_SECTION_SIZE = 1;
     private final PoiManager.DistanceTracker distanceTracker;
     private final LongSet loadedChunks = new LongOpenHashSet();
+    // DeerFolia - Async path processing
+    private final java.util.concurrent.locks.ReadWriteLock rwlock = new java.util.concurrent.locks.ReentrantReadWriteLock();
+    private final java.util.concurrent.locks.Lock rlock = this.rwlock.readLock();
+    private final java.util.concurrent.locks.Lock wlock = this.rwlock.writeLock();
+    // DeerFolia - Async path processing
 
     // Paper start - rewrite chunk system
     private final net.minecraft.server.level.ServerLevel world;
@@ -75,11 +80,24 @@ public class PoiManager extends SectionStorage<PoiSection, PoiSection.Packed> im
         final int chunkY = ca.spottedleaf.moonrise.common.util.CoordinateUtils.getChunkSectionY(pos);
         final int chunkZ = ca.spottedleaf.moonrise.common.util.CoordinateUtils.getChunkSectionZ(pos);
 
-        ca.spottedleaf.moonrise.common.util.TickThread.ensureTickThread(this.world, chunkX, chunkZ, "Accessing poi chunk off-main");
-
-        final ca.spottedleaf.moonrise.patches.chunk_system.level.poi.PoiChunk ret = ((ca.spottedleaf.moonrise.patches.chunk_system.level.ChunkSystemServerLevel)this.world).moonrise$getChunkTaskScheduler().chunkHolderManager.getPoiChunkIfLoaded(chunkX, chunkZ, true);
-
-        return ret == null ? Optional.empty() : ret.getSectionForVanilla(chunkY);
+        // DeerFolia - Async path processing
+        try {
+            if (this.rlock.tryLock(100, java.util.concurrent.TimeUnit.MILLISECONDS)) {
+                try {
+                    final ca.spottedleaf.moonrise.patches.chunk_system.level.poi.PoiChunk ret = ((ca.spottedleaf.moonrise.patches.chunk_system.level.ChunkSystemServerLevel)this.world)
+                        .moonrise$getChunkTaskScheduler()
+                        .chunkHolderManager
+                        .getPoiChunkIfLoaded(chunkX, chunkZ, true);
+                    return ret == null ? Optional.empty() : ret.getSectionForVanilla(chunkY);
+                } finally {
+                    this.rlock.unlock();
+                }
+            }
+            return Optional.empty();
+        } catch (InterruptedException e) {
+            return Optional.empty();
+        }
+        // DeerFolia - Async path processing
     }
 
     @Override
@@ -364,16 +382,40 @@ public class PoiManager extends SectionStorage<PoiSection, PoiSection.Packed> im
         final Predicate<Holder<PoiType>> predicate, final BiPredicate<Holder<PoiType>, BlockPos> filter, final BlockPos center, final int radius
     ) {
         // Paper start - optimise POI lookup
-        final PoiRecord record = ca.spottedleaf.moonrise.patches.poi_lookup.PoiAccess.findAnyPoiRecord(
-            (PoiManager)(Object)this, predicate, filter, center, radius, (double)((long)radius * (long)radius), PoiManager.Occupancy.HAS_SPACE, ca.spottedleaf.moonrise.patches.poi_lookup.PoiAccess.LOAD_FOR_SEARCHING
-        );
-
-        if (record == null) {
-            return Optional.empty();
+        // DeerFolia - Async path processing
+        try {
+            PoiRecord record = null;
+            if (this.rlock.tryLock(100, java.util.concurrent.TimeUnit.MILLISECONDS)) {
+                try {
+                    record = ca.spottedleaf.moonrise.patches.poi_lookup.PoiAccess.findAnyPoiRecord(
+                        (PoiManager)(Object)this,
+                        predicate,
+                        filter,
+                        center,
+                        radius,
+                        (double)((long)radius * (long)radius),
+                        PoiManager.Occupancy.HAS_SPACE,
+                        ca.spottedleaf.moonrise.patches.poi_lookup.PoiAccess.LOAD_FOR_SEARCHING
+                    );
+                } finally {
+                    this.rlock.unlock();
+                }
+            }
+            if (record == null) {
+                return Optional.empty();
+            }
+            if (this.wlock.tryLock(100, java.util.concurrent.TimeUnit.MILLISECONDS)) {
+                try {
+                    record.acquireTicket();
+                    return Optional.of(record.getPos());
+                } finally {
+                    this.wlock.unlock();
+                }
+            }
+        } catch (InterruptedException ignored) {
         }
-
-        record.acquireTicket();
-        return Optional.of(record.getPos());
+        return Optional.empty();
+        // DeerFolia - Async path processing
         // Paper end - optimise POI lookup
     }
 
@@ -412,6 +454,12 @@ public class PoiManager extends SectionStorage<PoiSection, PoiSection.Packed> im
         return this.getOrLoad(SectionPos.asLong(pos)).map(s -> s.exists(pos, predicate)).orElse(false);
     }
 
+    // DeerFolia - Async path processing
+    public Optional<Holder<PoiType>> getTypeWithoutLoad(final BlockPos pos) {
+        return this.get(SectionPos.asLong(pos)).flatMap(section -> section.getType(pos));
+    }
+    // DeerFolia - Async path processing
+
     public Optional<Holder<PoiType>> getType(final BlockPos pos) {
         return this.getOrLoad(SectionPos.asLong(pos)).flatMap(section -> section.getType(pos));
     }
diff --git a/net/minecraft/world/entity/animal/frog/Frog.java b/net/minecraft/world/entity/animal/frog/Frog.java
index dbc4071ec75f9618901d4ce2cc7a01577807fcdc..a1d9c812b9f5581cbb5e07906106f9ce2d01fbcc 100644
--- a/net/minecraft/world/entity/animal/frog/Frog.java
+++ b/net/minecraft/world/entity/animal/frog/Frog.java
@@ -386,6 +386,17 @@ public class Frog extends Animal {
             super(mob, level);
         }
 
+        // DeerFolia start - petal - async path processing
+        private static final cn.lunadeer.mc.deerfolia.async.path.NodeEvaluatorGenerator nodeEvaluatorGenerator = nodeEvaluatorFeatures -> {
+            Frog.FrogNodeEvaluator nodeEvaluator = new Frog.FrogNodeEvaluator(true);
+            nodeEvaluator.setCanPassDoors(nodeEvaluatorFeatures.canPassDoors());
+            nodeEvaluator.setCanFloat(nodeEvaluatorFeatures.canFloat());
+            nodeEvaluator.setCanWalkOverFences(nodeEvaluatorFeatures.canWalkOverFences());
+            nodeEvaluator.setCanOpenDoors(nodeEvaluatorFeatures.canOpenDoors());
+            return nodeEvaluator;
+        };
+        // DeerFolia end
+
         @Override
         public boolean canCutCorner(final PathType pathType) {
             return pathType != PathType.WATER_BORDER && super.canCutCorner(pathType);
@@ -394,6 +405,11 @@ public class Frog extends Animal {
         @Override
         protected PathFinder createPathFinder(final int maxVisitedNodes) {
             this.nodeEvaluator = new Frog.FrogNodeEvaluator(true);
+            // DeerFolia start - petal - async path processing
+            if (cn.lunadeer.mc.deerfolia.DeerFoliaConfiguration.asyncPathfinding.enabled) {
+                return new PathFinder(this.nodeEvaluator, maxVisitedNodes, nodeEvaluatorGenerator);
+            }
+            // DeerFolia end
             return new PathFinder(this.nodeEvaluator, maxVisitedNodes);
         }
     }
diff --git a/net/minecraft/world/entity/monster/Strider.java b/net/minecraft/world/entity/monster/Strider.java
index 0fbbe892217a5b894a2132439cc44e07fed78d8b..42126ac9b93c6bdc3568c6646ea4920b51aff6e0 100644
--- a/net/minecraft/world/entity/monster/Strider.java
+++ b/net/minecraft/world/entity/monster/Strider.java
@@ -525,9 +525,25 @@ public class Strider extends Animal implements ItemSteerable {
             super(mob, level);
         }
 
+        // DeerFolia start - petal - async path processing
+        private static final cn.lunadeer.mc.deerfolia.async.path.NodeEvaluatorGenerator nodeEvaluatorGenerator = nodeEvaluatorFeatures -> {
+            WalkNodeEvaluator nodeEvaluator = new WalkNodeEvaluator();
+            nodeEvaluator.setCanPassDoors(nodeEvaluatorFeatures.canPassDoors());
+            nodeEvaluator.setCanFloat(nodeEvaluatorFeatures.canFloat());
+            nodeEvaluator.setCanWalkOverFences(nodeEvaluatorFeatures.canWalkOverFences());
+            nodeEvaluator.setCanOpenDoors(nodeEvaluatorFeatures.canOpenDoors());
+            return nodeEvaluator;
+        };
+        // DeerFolia end
+
         @Override
         protected PathFinder createPathFinder(final int maxVisitedNodes) {
             this.nodeEvaluator = new WalkNodeEvaluator();
+            // DeerFolia start - async path processing
+            if (cn.lunadeer.mc.deerfolia.DeerFoliaConfiguration.asyncPathfinding.enabled) {
+                return new PathFinder(this.nodeEvaluator, maxVisitedNodes, nodeEvaluatorGenerator);
+            }
+            // DeerFolia end
             return new PathFinder(this.nodeEvaluator, maxVisitedNodes);
         }
 
diff --git a/net/minecraft/world/entity/monster/warden/Warden.java b/net/minecraft/world/entity/monster/warden/Warden.java
index 3b4567fe9bf95acef9499a5455e59fbe2ef29f05..cf02ff691395f27ea4b114df46f5f0c289cecd2d 100644
--- a/net/minecraft/world/entity/monster/warden/Warden.java
+++ b/net/minecraft/world/entity/monster/warden/Warden.java
@@ -561,6 +561,16 @@ public class Warden extends Monster implements VibrationSystem {
             @Override
             protected PathFinder createPathFinder(final int maxVisitedNodes) {
                 this.nodeEvaluator = new WalkNodeEvaluator();
+                // DeerFolia start - petal - async path processing
+                if (cn.lunadeer.mc.deerfolia.DeerFoliaConfiguration.asyncPathfinding.enabled) {
+                    return new PathFinder(this.nodeEvaluator, maxVisitedNodes, GroundPathNavigation.nodeEvaluatorGenerator) {
+                        @Override
+                        protected float distance(final Node from, final Node to) {
+                            return from.distanceToXZ(to);
+                        }
+                    };
+                }
+                // DeerFolia end
                 return new PathFinder(this.nodeEvaluator, maxVisitedNodes) {
                     @Override
                     protected float distance(final Node from, final Node to) {
diff --git a/net/minecraft/world/level/chunk/LevelChunk.java b/net/minecraft/world/level/chunk/LevelChunk.java
index 64e0c4a60a74cd33f2b99f34fa2a419abf3a0c02..47b542e167f2d4d049b60c7ef218aa5d4f7749e3 100644
--- a/net/minecraft/world/level/chunk/LevelChunk.java
+++ b/net/minecraft/world/level/chunk/LevelChunk.java
@@ -472,7 +472,10 @@ public class LevelChunk extends ChunkAccess implements DebugValueSource, ca.spot
 
     public @Nullable BlockEntity getBlockEntity(final BlockPos pos, final LevelChunk.EntityCreationType creationType) {
         // CraftBukkit start
-        BlockEntity blockEntity = this.level.getCurrentWorldData().capturedTileEntities.get(pos); // Folia - region threading
+        // BlockEntity blockEntity = this.level.getCurrentWorldData().capturedTileEntities.get(pos); // Folia - region threading
+        BlockEntity blockEntity = java.util.Optional.ofNullable(this.level.getCurrentWorldData())
+            .map(data -> data.capturedTileEntities.get(pos))
+            .orElse(null);
         if (blockEntity == null) {
             blockEntity = this.blockEntities.get(pos);
         }
diff --git a/net/minecraft/world/level/pathfinder/Path.java b/net/minecraft/world/level/pathfinder/Path.java
index ce536d07512c3cb9a8f56a1c4f6506db50161f8b..4aaaa67d9b64f1795224bd6be9ccf072a4ca1ce5 100644
--- a/net/minecraft/world/level/pathfinder/Path.java
+++ b/net/minecraft/world/level/pathfinder/Path.java
@@ -11,7 +11,7 @@ import net.minecraft.world.entity.Entity;
 import net.minecraft.world.phys.Vec3;
 import org.jspecify.annotations.Nullable;
 
-public final class Path {
+public class Path {
     public static final StreamCodec<FriendlyByteBuf, Path> STREAM_CODEC = StreamCodec.of((output, value) -> value.writeToStream(output), Path::createFromStream);
     public final List<Node> nodes;
     private Path.@Nullable DebugData debugData;
@@ -27,6 +27,12 @@ public final class Path {
         this.reached = reached;
     }
 
+    // DeerFolia start - petal - async path processing
+    public boolean isProcessed() {
+        return true;
+    }
+    // DeerFolia end
+
     public void advance() {
         this.nextNodeIndex++;
     }
@@ -98,6 +104,7 @@ public final class Path {
     }
 
     public boolean sameAs(final @Nullable Path path) {
+        if (path == this) return true; // DeerFolia - petal - short circuit
         return path != null && this.nodes.equals(path.nodes);
     }
 
diff --git a/net/minecraft/world/level/pathfinder/PathFinder.java b/net/minecraft/world/level/pathfinder/PathFinder.java
index 253074a0a14115cd3d18aaae6fa5822cf32346c9..1ab4043dc74a9407dc57c2bed78aa631719feb06 100644
--- a/net/minecraft/world/level/pathfinder/PathFinder.java
+++ b/net/minecraft/world/level/pathfinder/PathFinder.java
@@ -26,10 +26,22 @@ public class PathFinder {
     public final NodeEvaluator nodeEvaluator;
     private final BinaryHeap openSet = new BinaryHeap();
     private BooleanSupplier captureDebug = () -> false;
+    private final cn.lunadeer.mc.deerfolia.async.path.NodeEvaluatorGenerator nodeEvaluatorGenerator; // DeerFolia - petal - we use this later to generate an evaluator
 
-    public PathFinder(final NodeEvaluator nodeEvaluator, final int maxVisitedNodes) {
+    public PathFinder(
+        final NodeEvaluator nodeEvaluator,
+        final int maxVisitedNodes,
+        final cn.lunadeer.mc.deerfolia.async.path.NodeEvaluatorGenerator nodeEvaluatorGenerator
+    ) { // DeerFolia - petal - add nodeEvaluatorGenerator
         this.nodeEvaluator = nodeEvaluator;
         this.maxVisitedNodes = maxVisitedNodes;
+        // DeerFolia start - petal - support nodeEvaluatorgenerators
+        this.nodeEvaluatorGenerator = nodeEvaluatorGenerator;
+    }
+
+    public PathFinder(final NodeEvaluator nodeEvaluator, final int maxVisitedNodes) {
+        this(nodeEvaluator, maxVisitedNodes, null);
+        // DeerFolia end
     }
 
     public void setCaptureDebug(final BooleanSupplier captureDebug) {
@@ -48,27 +60,69 @@ public class PathFinder {
         final int reachRange,
         final float maxVisitedNodesMultiplier
     ) {
-        this.openSet.clear();
-        this.nodeEvaluator.prepare(level, entity);
-        Node from = this.nodeEvaluator.getStart();
+        if (!cn.lunadeer.mc.deerfolia.DeerFoliaConfiguration.asyncPathfinding.enabled)
+            this.openSet.clear(); // DeerFolia - petal - it's always cleared in processPath
+        // DeerFolia start - petal - use a generated evaluator if we have one otherwise run sync
+        NodeEvaluator nodeEvaluator = this.nodeEvaluatorGenerator == null
+            ? this.nodeEvaluator
+            : cn.lunadeer.mc.deerfolia.async.path.NodeEvaluatorCache.takeNodeEvaluator(this.nodeEvaluatorGenerator, this.nodeEvaluator);
+        nodeEvaluator.prepare(level, entity);
+        Node from = nodeEvaluator.getStart();
+        // DeerFolia end
         if (from == null) {
+            cn.lunadeer.mc.deerfolia.async.path.NodeEvaluatorCache.removeNodeEvaluator(nodeEvaluator); // DeerFolia - petal - handle nodeEvaluatorGenerator
             return null;
         }
 
         // Paper start - Perf: remove streams and optimize collection
         List<Map.Entry<Target, BlockPos>> tos = Lists.newArrayList();
         for (BlockPos pos : targets) {
-            tos.add(new java.util.AbstractMap.SimpleEntry<>(this.nodeEvaluator.getTarget(pos.getX(), pos.getY(), pos.getZ()), pos));
+            tos.add(new java.util.AbstractMap.SimpleEntry<>(nodeEvaluator.getTarget(pos.getX(), pos.getY(), pos.getZ()), pos)); // DeerFolia - petal - handle nodeEvaluatorGenerator
         }
         // Paper end - Perf: remove streams and optimize collection
-        Path path = this.findPath(from, tos, maxPathLength, reachRange, maxVisitedNodesMultiplier);
-        this.nodeEvaluator.done();
-        return path;
+        // DeerFolia start - petal - async path processing
+        if (this.nodeEvaluatorGenerator == null) {
+            cn.lunadeer.mc.deerfolia.async.path.NodeEvaluatorCache.removeNodeEvaluator(nodeEvaluator);
+            return this.findPath(from, tos, maxPathLength, reachRange, maxVisitedNodesMultiplier);
+        }
+        return new cn.lunadeer.mc.deerfolia.async.path.AsyncPath(Lists.newArrayList(), targets, () -> {
+            try {
+                return this.processPath(nodeEvaluator, from, tos, maxPathLength, reachRange, maxVisitedNodesMultiplier);
+            } catch (Exception e) {
+                e.printStackTrace();
+                return null;
+            } finally {
+                nodeEvaluator.done();
+                cn.lunadeer.mc.deerfolia.async.path.NodeEvaluatorCache.returnNodeEvaluator(nodeEvaluator);
+            }
+        });
+        // DeerFolia end
     }
 
     private @Nullable Path findPath(
         final Node from, final List<Map.Entry<Target, BlockPos>> targets, final float maxPathLength, final int reachRange, final float maxVisitedNodesMultiplier // Paper - optimize collection
     ) {
+        // DeerFolia start - petal - split pathfinding into the original sync method for compat and processing for delaying
+        try {
+            return this.processPath(this.nodeEvaluator, from, targets, maxPathLength, reachRange, maxVisitedNodesMultiplier);
+        } catch (Exception e) {
+            e.printStackTrace();
+            return null;
+        } finally {
+            this.nodeEvaluator.done();
+        }
+    }
+
+    private synchronized @org.jetbrains.annotations.NotNull Path processPath(
+        final NodeEvaluator nodeEvaluator,
+        final Node from,
+        final List<Map.Entry<Target, BlockPos>> targets,
+        final float maxPathLength,
+        final int reachRange,
+        final float maxVisitedNodesMultiplier
+    ) { // sync to only use the caching functions in this class on a single thread
+        org.apache.commons.lang3.Validate.isTrue(!targets.isEmpty()); // ensure that we have at least one position, which means we'll always return a path
+        // DeerFolia end
         ProfilerFiller profiler = Profiler.get();
         profiler.push("find_path");
         profiler.markForCharting(MetricCategory.PATH_FINDING);
@@ -112,7 +166,7 @@ public class PathFinder {
             }
 
             if (!(current.distanceTo(from) >= maxPathLength)) {
-                int neighborCount = this.nodeEvaluator.getNeighbors(this.neighbors, current);
+                int neighborCount = nodeEvaluator.getNeighbors(this.neighbors, current); // DeerFolia - petal - use provided nodeEvaluator
 
                 for (int i = 0; i < neighborCount; i++) {
                     Node neighbor = this.neighbors[i];
diff --git a/net/minecraft/world/level/pathfinder/SwimNodeEvaluator.java b/net/minecraft/world/level/pathfinder/SwimNodeEvaluator.java
index e55558448f91410f6174548edc88327191358825..eaf6f24f5a1088a012a1617f9053c999d4bce35b 100644
--- a/net/minecraft/world/level/pathfinder/SwimNodeEvaluator.java
+++ b/net/minecraft/world/level/pathfinder/SwimNodeEvaluator.java
@@ -15,7 +15,7 @@ import net.minecraft.world.level.material.FluidState;
 import org.jspecify.annotations.Nullable;
 
 public class SwimNodeEvaluator extends NodeEvaluator {
-    private final boolean allowBreaching;
+    public final boolean allowBreaching; // DeerFolia - make this public
     private final Long2ObjectMap<PathType> pathTypesByPosCache = new Long2ObjectOpenHashMap<>();
 
     public SwimNodeEvaluator(final boolean allowBreaching) {
