From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: zhangyuheng <zhangyuheng@lunadeer.cn>
Date: Tue, 11 Mar 2025 18:04:31 +0800
Subject: [PATCH] Async pathfinding


diff --git a/net/minecraft/world/entity/ai/behavior/AcquirePoi.java b/net/minecraft/world/entity/ai/behavior/AcquirePoi.java
index a8f2b22436ff0f4bff06b2c5cf01375da755cfde..758b69aeff096183fa07d43e5b34ec006055a56d 100644
--- a/net/minecraft/world/entity/ai/behavior/AcquirePoi.java
+++ b/net/minecraft/world/entity/ai/behavior/AcquirePoi.java
@@ -98,6 +98,30 @@ public class AcquirePoi {
                                     }
                                 }
                                 // Paper end - optimise POI searches
+                                // DeerFolia start - petal - Async path processing
+                                if (cn.lunadeer.mc.deerfolia.DeerFoliaConfiguration.asyncPathfinding.enabled) {
+                                    Path possiblePath = findPathToPois(body, poiPositions);
+                                    cn.lunadeer.mc.deerfolia.async.path.AsyncPathProcessor.awaitProcessing(body, possiblePath, path -> {
+                                        if (path == null || !path.canReach()) {
+                                            for (Pair<Holder<PoiType>, BlockPos> p : poiPositions) {
+                                                batchCache.computeIfAbsent(
+                                                    p.getSecond().asLong(), key -> new AcquirePoi.JitteredLinearRetry(random, timestamp)
+                                                );
+                                            }
+                                            return;
+                                        }
+
+                                        BlockPos targetPos = path.getTarget();
+                                        poiManager.getTypeWithoutLoad(targetPos).ifPresent(type -> {
+                                            poiManager.take(poiType, (t, poiPos) -> poiPos.equals(targetPos), targetPos, 1);
+                                            toAcquire.set(GlobalPos.of(level.dimension(), targetPos));
+                                            onPoiAcquisitionEvent.ifPresent(event -> level.broadcastEntityEvent(body, event));
+                                            batchCache.clear();
+                                            level.debugSynchronizers().updatePoi(targetPos);
+                                        });
+                                    });
+                                } else {
+                                // DeerFolia end
                                 Path path = findPathToPois(body, poiPositions);
                                 if (path != null && path.canReach()) {
                                     BlockPos targetPos = path.getTarget();
@@ -113,6 +137,7 @@ public class AcquirePoi {
                                         batchCache.computeIfAbsent(p.getSecond().asLong(), key -> new AcquirePoi.JitteredLinearRetry(random, timestamp));
                                     }
                                 }
+                                } // DeerFolia - async path processing
 
                                 return true;
                             }
diff --git a/net/minecraft/world/entity/ai/behavior/MoveToTargetSink.java b/net/minecraft/world/entity/ai/behavior/MoveToTargetSink.java
index 194813475271364831cdd1bb5ae244967341f9da..7f35d5142c26192a8244b9f15dfb24156a62ec37 100644
--- a/net/minecraft/world/entity/ai/behavior/MoveToTargetSink.java
+++ b/net/minecraft/world/entity/ai/behavior/MoveToTargetSink.java
@@ -21,6 +21,7 @@ public class MoveToTargetSink extends Behavior<Mob> {
     private int remainingCooldown;
     private @Nullable Path path;
     private @Nullable BlockPos lastTargetPos;
+    private boolean finishedProcessing; // DeerFolia - petal - track when path is processed
     private float speedModifier;
 
     public MoveToTargetSink() {
@@ -51,9 +52,13 @@ public class MoveToTargetSink extends Behavior<Mob> {
             Brain<?> brain = body.getBrain();
             WalkTarget walkTarget = brain.getMemory(MemoryModuleType.WALK_TARGET).get();
             boolean reachedTarget = this.reachedTarget(body, walkTarget);
-            if (!reachedTarget && this.tryComputePath(body, walkTarget, level.getGameTime())) {
+            if (!cn.lunadeer.mc.deerfolia.DeerFoliaConfiguration.asyncPathfinding.enabled
+                && !reachedTarget
+                && this.tryComputePath(body, walkTarget, level.getGameTime())) { // DeerFolia - petal - async path processing means we can't know if the path is reachable here
                 this.lastTargetPos = walkTarget.getTarget().currentBlockPosition();
                 return true;
+            } else if (cn.lunadeer.mc.deerfolia.DeerFoliaConfiguration.asyncPathfinding.enabled && !reachedTarget) {
+                return true; // DeerFolia - async pathfinding
             } else {
                 brain.eraseMemory(MemoryModuleType.WALK_TARGET);
                 if (reachedTarget) {
@@ -67,6 +72,7 @@ public class MoveToTargetSink extends Behavior<Mob> {
 
     @Override
     protected boolean canStillUse(final ServerLevel level, final Mob body, final long timestamp) {
+        if (cn.lunadeer.mc.deerfolia.DeerFoliaConfiguration.asyncPathfinding.enabled && !this.finishedProcessing) return true; // DeerFolia - petal - wait for processing
         if (this.path != null && this.lastTargetPos != null) {
             Optional<WalkTarget> walkTarget = body.getBrain().getMemory(MemoryModuleType.WALK_TARGET);
             boolean isSpectator = walkTarget.map(MoveToTargetSink::isWalkTargetSpectator).orElse(false);
@@ -93,27 +99,97 @@ public class MoveToTargetSink extends Behavior<Mob> {
 
     @Override
     protected void start(final ServerLevel level, final Mob body, final long timestamp) {
+        // DeerFolia start - petal - start processing
+        if (cn.lunadeer.mc.deerfolia.DeerFoliaConfiguration.asyncPathfinding.enabled) {
+            Brain<?> brain = body.getBrain();
+            WalkTarget walkTarget = brain.getMemory(MemoryModuleType.WALK_TARGET).get();
+
+            this.finishedProcessing = false;
+            this.lastTargetPos = walkTarget.getTarget().currentBlockPosition();
+            this.path = this.computePath(body, walkTarget);
+            return;
+        }
+        // DeerFolia end
         body.getBrain().setMemory(MemoryModuleType.PATH, this.path);
         body.getNavigation().moveTo(this.path, (double)this.speedModifier);
     }
 
     @Override
     protected void tick(final ServerLevel level, final Mob body, final long timestamp) {
-        Path newPath = body.getNavigation().getPath();
-        Brain<?> brain = body.getBrain();
-        if (this.path != newPath) {
-            this.path = newPath;
-            brain.setMemory(MemoryModuleType.PATH, newPath);
-        }
+        // DeerFolia start - petal - Async path processing
+        if (cn.lunadeer.mc.deerfolia.DeerFoliaConfiguration.asyncPathfinding.enabled) {
+            if (this.path != null && !this.path.isProcessed()) return; // wait for processing
 
-        if (newPath != null && this.lastTargetPos != null) {
-            WalkTarget walkTarget = brain.getMemory(MemoryModuleType.WALK_TARGET).get();
-            if (walkTarget.getTarget().currentBlockPosition().distSqr(this.lastTargetPos) > 4.0 && this.tryComputePath(body, walkTarget, level.getGameTime())) {
-                this.lastTargetPos = walkTarget.getTarget().currentBlockPosition();
-                this.start(level, body, timestamp);
+            if (!this.finishedProcessing) {
+                this.finishedProcessing = true;
+
+                Brain<?> brain = body.getBrain();
+                boolean canReach = this.path != null && this.path.canReach();
+                if (canReach) {
+                    brain.eraseMemory(MemoryModuleType.CANT_REACH_WALK_TARGET_SINCE);
+                } else if (!brain.hasMemoryValue(MemoryModuleType.CANT_REACH_WALK_TARGET_SINCE)) {
+                    brain.setMemory(MemoryModuleType.CANT_REACH_WALK_TARGET_SINCE, timestamp);
+                }
+
+                if (!canReach) {
+                    Optional<WalkTarget> walkTarget = brain.getMemory(MemoryModuleType.WALK_TARGET);
+                    if (walkTarget.isEmpty()) {
+                        return;
+                    }
+
+                    BlockPos targetPos = walkTarget.get().getTarget().currentBlockPosition();
+                    Vec3 partialStep = DefaultRandomPos.getPosTowards((PathfinderMob)body, 10, 7, Vec3.atBottomCenterOf(targetPos), (float)(Math.PI / 2));
+                    if (partialStep != null) {
+                        this.path = body.getNavigation().createPath(partialStep.x, partialStep.y, partialStep.z, 0);
+                        this.finishedProcessing = false;
+                        return;
+                    }
+                }
+
+                body.getBrain().setMemory(MemoryModuleType.PATH, this.path);
+                body.getNavigation().moveTo(this.path, this.speedModifier);
+            }
+
+            Path path = body.getNavigation().getPath();
+            Brain<?> brain = body.getBrain();
+            if (path != null && this.lastTargetPos != null && brain.hasMemoryValue(MemoryModuleType.WALK_TARGET)) {
+                WalkTarget walkTarget = brain.getMemory(MemoryModuleType.WALK_TARGET).get();
+                if (walkTarget.getTarget().currentBlockPosition().distSqr(this.lastTargetPos) > 4.0) {
+                    this.start(level, body, timestamp);
+                }
+            }
+        } else {
+        // DeerFolia end
+            Path newPath = body.getNavigation().getPath();
+            Brain<?> brain = body.getBrain();
+            if (this.path != newPath) {
+                this.path = newPath;
+                brain.setMemory(MemoryModuleType.PATH, newPath);
+            }
+
+            if (newPath != null && this.lastTargetPos != null) {
+                WalkTarget walkTarget = brain.getMemory(MemoryModuleType.WALK_TARGET).get();
+                if (walkTarget.getTarget().currentBlockPosition().distSqr(this.lastTargetPos) > 4.0
+                    && this.tryComputePath(body, walkTarget, level.getGameTime())) {
+                    this.lastTargetPos = walkTarget.getTarget().currentBlockPosition();
+                    this.start(level, body, timestamp);
+                }
             }
+        } // DeerFolia - async path processing
+    }
+
+    // DeerFolia start - petal - Async path processing
+    private @Nullable Path computePath(final Mob body, final WalkTarget walkTarget) {
+        BlockPos targetPos = walkTarget.getTarget().currentBlockPosition();
+        this.speedModifier = walkTarget.getSpeedModifier();
+        Brain<?> brain = body.getBrain();
+        if (this.reachedTarget(body, walkTarget)) {
+            brain.eraseMemory(MemoryModuleType.CANT_REACH_WALK_TARGET_SINCE);
         }
+
+        return body.getNavigation().createPath(targetPos, 0);
     }
+    // DeerFolia end
 
     private boolean tryComputePath(final Mob body, final WalkTarget walkTarget, final long timestamp) {
         BlockPos targetPos = walkTarget.getTarget().currentBlockPosition();
diff --git a/net/minecraft/world/entity/ai/behavior/SetClosestHomeAsWalkTarget.java b/net/minecraft/world/entity/ai/behavior/SetClosestHomeAsWalkTarget.java
index 843d6897a43ed9e9d28b29608401b78266f57f61..d570c5e6930ab898847556bf213e3b132785d719 100644
--- a/net/minecraft/world/entity/ai/behavior/SetClosestHomeAsWalkTarget.java
+++ b/net/minecraft/world/entity/ai/behavior/SetClosestHomeAsWalkTarget.java
@@ -59,6 +59,27 @@ public class SetClosestHomeAsWalkTarget {
                                         p -> p.is(PoiTypes.HOME), cacheTest, body.blockPosition(), AcquirePoi.SCAN_RANGE, PoiManager.Occupancy.ANY
                                     )
                                     .collect(Collectors.toSet());
+                                // DeerFolia start - petal - Async path processing
+                                if (cn.lunadeer.mc.deerfolia.DeerFoliaConfiguration.asyncPathfinding.enabled) {
+                                    Path possiblePath = AcquirePoi.findPathToPois(body, pois);
+                                    cn.lunadeer.mc.deerfolia.async.path.AsyncPathProcessor.awaitProcessing(body, possiblePath, path -> {
+                                        if ((path == null || !path.canReach()) && triedCount.intValue() < BATCH_SIZE) {
+                                            batchCache.long2LongEntrySet().removeIf(entry -> entry.getLongValue() < lastUpdate.longValue());
+                                            return;
+                                        }
+                                        if (path == null) {
+                                            return;
+                                        }
+
+                                        BlockPos targetPos = path.getTarget();
+                                        Optional<Holder<PoiType>> type = poiManager.getTypeWithoutLoad(targetPos);
+                                        if (type.isPresent()) {
+                                            walkTarget.set(new WalkTarget(targetPos, speedModifier, 1));
+                                            level.debugSynchronizers().updatePoi(targetPos);
+                                        }
+                                    });
+                                } else {
+                                // DeerFolia end
                                 Path path = AcquirePoi.findPathToPois(body, pois);
                                 if (path != null && path.canReach()) {
                                     BlockPos targetPos = path.getTarget();
@@ -70,6 +91,7 @@ public class SetClosestHomeAsWalkTarget {
                                 } else if (triedCount.intValue() < BATCH_SIZE) {
                                     batchCache.long2LongEntrySet().removeIf(entry -> entry.getLongValue() < lastUpdate.longValue());
                                 }
+                                } // DeerFolia - async path processing
 
                                 return true;
                             } else {
diff --git a/net/minecraft/world/entity/ai/goal/DoorInteractGoal.java b/net/minecraft/world/entity/ai/goal/DoorInteractGoal.java
index b45807e66f927344cdaccf0727c5e2e2a96eb460..42f98b35999f10e15c679f6af4e51872323ca5d7 100644
--- a/net/minecraft/world/entity/ai/goal/DoorInteractGoal.java
+++ b/net/minecraft/world/entity/ai/goal/DoorInteractGoal.java
@@ -54,7 +54,7 @@ public abstract class DoorInteractGoal extends Goal {
             return false;
         } else {
             Path path = this.mob.getNavigation().getPath();
-            if (path != null && !path.isDone()) {
+            if (path != null && path.isProcessed() && !path.isDone()) { // DeerFolia - async pathfinding - ensure path is processed
                 for (int i = 0; i < Math.min(path.getNextNodeIndex() + 2, path.getNodeCount()); i++) {
                     Node node = path.getNode(i);
                     this.doorPos = new BlockPos(node.x, node.y + 1, node.z);
diff --git a/net/minecraft/world/entity/ai/navigation/AmphibiousPathNavigation.java b/net/minecraft/world/entity/ai/navigation/AmphibiousPathNavigation.java
index 45942481a972f2308eae80f6f0f8b886e47016ec..b5c726bc060e00593d49604a49009b2c27f74d2c 100644
--- a/net/minecraft/world/entity/ai/navigation/AmphibiousPathNavigation.java
+++ b/net/minecraft/world/entity/ai/navigation/AmphibiousPathNavigation.java
@@ -12,9 +12,25 @@ public class AmphibiousPathNavigation extends PathNavigation {
         super(mob, level);
     }
 
+    // DeerFolia start - petal - async path processing
+    private static final cn.lunadeer.mc.deerfolia.async.path.NodeEvaluatorGenerator nodeEvaluatorGenerator = nodeEvaluatorFeatures -> {
+        AmphibiousNodeEvaluator nodeEvaluator = new AmphibiousNodeEvaluator(false);
+        nodeEvaluator.setCanPassDoors(nodeEvaluatorFeatures.canPassDoors());
+        nodeEvaluator.setCanFloat(nodeEvaluatorFeatures.canFloat());
+        nodeEvaluator.setCanWalkOverFences(nodeEvaluatorFeatures.canWalkOverFences());
+        nodeEvaluator.setCanOpenDoors(nodeEvaluatorFeatures.canOpenDoors());
+        return nodeEvaluator;
+    };
+    // DeerFolia end
+
     @Override
     protected PathFinder createPathFinder(final int maxVisitedNodes) {
         this.nodeEvaluator = new AmphibiousNodeEvaluator(false);
+        // DeerFolia start - petal - async path processing
+        if (cn.lunadeer.mc.deerfolia.DeerFoliaConfiguration.asyncPathfinding.enabled) {
+            return new PathFinder(this.nodeEvaluator, maxVisitedNodes, nodeEvaluatorGenerator);
+        }
+        // DeerFolia end
         return new PathFinder(this.nodeEvaluator, maxVisitedNodes);
     }
 
diff --git a/net/minecraft/world/entity/ai/navigation/FlyingPathNavigation.java b/net/minecraft/world/entity/ai/navigation/FlyingPathNavigation.java
index e44814cfb6afb594456b8215bd13a92e93de2c85..a87622eb56a224eeeec26351ca4890b000816f94 100644
--- a/net/minecraft/world/entity/ai/navigation/FlyingPathNavigation.java
+++ b/net/minecraft/world/entity/ai/navigation/FlyingPathNavigation.java
@@ -15,9 +15,25 @@ public class FlyingPathNavigation extends PathNavigation {
         super(mob, level);
     }
 
+    // DeerFolia start - petal - async path processing
+    private static final cn.lunadeer.mc.deerfolia.async.path.NodeEvaluatorGenerator nodeEvaluatorGenerator = nodeEvaluatorFeatures -> {
+        FlyNodeEvaluator nodeEvaluator = new FlyNodeEvaluator();
+        nodeEvaluator.setCanPassDoors(nodeEvaluatorFeatures.canPassDoors());
+        nodeEvaluator.setCanFloat(nodeEvaluatorFeatures.canFloat());
+        nodeEvaluator.setCanWalkOverFences(nodeEvaluatorFeatures.canWalkOverFences());
+        nodeEvaluator.setCanOpenDoors(nodeEvaluatorFeatures.canOpenDoors());
+        return nodeEvaluator;
+    };
+    // DeerFolia end
+
     @Override
     protected PathFinder createPathFinder(final int maxVisitedNodes) {
         this.nodeEvaluator = new FlyNodeEvaluator();
+        // DeerFolia start - petal - async path processing
+        if (cn.lunadeer.mc.deerfolia.DeerFoliaConfiguration.asyncPathfinding.enabled) {
+            return new PathFinder(this.nodeEvaluator, maxVisitedNodes, nodeEvaluatorGenerator);
+        }
+        // DeerFolia end
         return new PathFinder(this.nodeEvaluator, maxVisitedNodes);
     }
 
@@ -47,6 +63,7 @@ public class FlyingPathNavigation extends PathNavigation {
         if (this.hasDelayedRecomputation) {
             this.recomputePath();
         }
+        if (this.path != null && !this.path.isProcessed()) return; // DeerFolia - petal - async path processing
 
         if (!this.isDone()) {
             if (this.canUpdatePath()) {
diff --git a/net/minecraft/world/entity/ai/navigation/GroundPathNavigation.java b/net/minecraft/world/entity/ai/navigation/GroundPathNavigation.java
index 85b694b110ed85992c87c1268de7124c78a609b7..ff1b9b1a16b142021d79cd204a152b072325cd85 100644
--- a/net/minecraft/world/entity/ai/navigation/GroundPathNavigation.java
+++ b/net/minecraft/world/entity/ai/navigation/GroundPathNavigation.java
@@ -25,9 +25,25 @@ public class GroundPathNavigation extends PathNavigation {
         super(mob, level);
     }
 
+    // DeerFolia start - petal - async path processing
+    protected static final cn.lunadeer.mc.deerfolia.async.path.NodeEvaluatorGenerator nodeEvaluatorGenerator = nodeEvaluatorFeatures -> {
+        WalkNodeEvaluator nodeEvaluator = new WalkNodeEvaluator();
+        nodeEvaluator.setCanPassDoors(nodeEvaluatorFeatures.canPassDoors());
+        nodeEvaluator.setCanFloat(nodeEvaluatorFeatures.canFloat());
+        nodeEvaluator.setCanWalkOverFences(nodeEvaluatorFeatures.canWalkOverFences());
+        nodeEvaluator.setCanOpenDoors(nodeEvaluatorFeatures.canOpenDoors());
+        return nodeEvaluator;
+    };
+    // DeerFolia end
+
     @Override
     protected PathFinder createPathFinder(final int maxVisitedNodes) {
         this.nodeEvaluator = new WalkNodeEvaluator();
+        // DeerFolia start - petal - async path processing
+        if (cn.lunadeer.mc.deerfolia.DeerFoliaConfiguration.asyncPathfinding.enabled) {
+            return new PathFinder(this.nodeEvaluator, maxVisitedNodes, nodeEvaluatorGenerator);
+        }
+        // DeerFolia end
         return new PathFinder(this.nodeEvaluator, maxVisitedNodes);
     }
 
diff --git a/net/minecraft/world/entity/ai/navigation/PathNavigation.java b/net/minecraft/world/entity/ai/navigation/PathNavigation.java
index 21ead01c326b187dc1117b549dbdf706730a2147..75365501dad75af2764fcfd6e464657cf09e8089 100644
--- a/net/minecraft/world/entity/ai/navigation/PathNavigation.java
+++ b/net/minecraft/world/entity/ai/navigation/PathNavigation.java
@@ -161,6 +161,12 @@ public abstract class PathNavigation {
             return null;
         } else if (!this.canUpdatePath()) {
             return null;
+        // DeerFolia start - petal - catch early if it's still processing these positions let it keep processing
+        } else if (this.path instanceof cn.lunadeer.mc.deerfolia.async.path.AsyncPath asyncPath
+            && !asyncPath.isProcessed()
+            && asyncPath.hasSameProcessingPositions(targets)) {
+            return this.path;
+        // DeerFolia end
         } else if (this.path != null && !this.path.isDone() && targets.contains(this.targetPos)) {
             return this.path;
         } else {
@@ -190,11 +196,29 @@ public abstract class PathNavigation {
             );
             Path path = this.pathFinder.findPath(region, this.mob, targets, maxPathLength, reachRange, this.maxVisitedNodesMultiplier);
             profiler.pop();
+            // DeerFolia start - petal - async path processing
+            if (cn.lunadeer.mc.deerfolia.DeerFoliaConfiguration.asyncPathfinding.enabled) {
+                if (!targets.isEmpty()) {
+                    this.targetPos = targets.iterator().next();
+                }
+                cn.lunadeer.mc.deerfolia.async.path.AsyncPathProcessor.awaitProcessing(this.mob, path, processedPath -> {
+                    if (processedPath != this.path) {
+                        return;
+                    }
+                    if (processedPath != null && processedPath.getTarget() != null) {
+                        this.targetPos = processedPath.getTarget();
+                        this.reachRange = reachRange;
+                        this.resetStuckTimeout();
+                    }
+                });
+            } else {
+            // DeerFolia end
             if (path != null && path.getTarget() != null) {
                 this.targetPos = path.getTarget();
                 this.reachRange = reachRange;
                 this.resetStuckTimeout();
             }
+            } // DeerFolia - async path processing
 
             return path;
         }
@@ -245,8 +269,8 @@ public abstract class PathNavigation {
             if (this.isDone()) {
                 return false;
             } else {
-                this.trimPath();
-                if (this.path.getNodeCount() <= 0) {
+                if (this.path.isProcessed()) this.trimPath(); // DeerFolia - petal - only trim if processed
+                if (this.path.isProcessed() && this.path.getNodeCount() <= 0) { // DeerFolia - petal - only check node count if processed
                     return false;
                 } else {
                     this.speedModifier = speedModifier;
@@ -268,6 +292,7 @@ public abstract class PathNavigation {
         if (this.hasDelayedRecomputation) {
             this.recomputePath();
         }
+        if (this.path != null && !this.path.isProcessed()) return; // DeerFolia - petal - skip pathfinding if we're still processing
 
         if (!this.isDone()) {
             if (this.canUpdatePath()) {
@@ -293,6 +318,7 @@ public abstract class PathNavigation {
     }
 
     protected void followThePath() {
+        if (!this.path.isProcessed()) return; // DeerFolia - petal - skip if not processed
         Vec3 mobPos = this.getTempMobPos();
         this.maxDistanceToWaypoint = this.mob.getBbWidth() > 0.75F ? this.mob.getBbWidth() / 2.0F : 0.75F - this.mob.getBbWidth() / 2.0F;
         Vec3i currentNodePos = this.path.getNextNodePos();
@@ -449,7 +475,7 @@ public abstract class PathNavigation {
     public boolean shouldRecomputePath(final BlockPos pos) {
         if (this.hasDelayedRecomputation) {
             return false;
-        } else if (this.path != null && !this.path.isDone() && this.path.getNodeCount() != 0) {
+        } else if (this.path != null && this.path.isProcessed() && !this.path.isDone() && this.path.getNodeCount() != 0) { // DeerFolia - petal - Skip if not processed
             Node target = this.path.getEndNode();
             Vec3 middlePos = new Vec3((target.x + this.mob.getX()) / 2.0, (target.y + this.mob.getY()) / 2.0, (target.z + this.mob.getZ()) / 2.0);
             return pos.closerToCenterThan(middlePos, this.path.getNodeCount() - this.path.getNextNodeIndex());
diff --git a/net/minecraft/world/entity/ai/navigation/WaterBoundPathNavigation.java b/net/minecraft/world/entity/ai/navigation/WaterBoundPathNavigation.java
index bf7c7dfc9dabe905490a9610b4201a0023ef69b3..f6f004683abbb78449c6dc92e66956b603fc9be7 100644
--- a/net/minecraft/world/entity/ai/navigation/WaterBoundPathNavigation.java
+++ b/net/minecraft/world/entity/ai/navigation/WaterBoundPathNavigation.java
@@ -15,11 +15,27 @@ public class WaterBoundPathNavigation extends PathNavigation {
         super(mob, level);
     }
 
+    // DeerFolia start - petal - async path processing
+    private static final cn.lunadeer.mc.deerfolia.async.path.NodeEvaluatorGenerator nodeEvaluatorGenerator = nodeEvaluatorFeatures -> {
+        SwimNodeEvaluator nodeEvaluator = new SwimNodeEvaluator(nodeEvaluatorFeatures.allowBreaching());
+        nodeEvaluator.setCanPassDoors(nodeEvaluatorFeatures.canPassDoors());
+        nodeEvaluator.setCanFloat(nodeEvaluatorFeatures.canFloat());
+        nodeEvaluator.setCanWalkOverFences(nodeEvaluatorFeatures.canWalkOverFences());
+        nodeEvaluator.setCanOpenDoors(nodeEvaluatorFeatures.canOpenDoors());
+        return nodeEvaluator;
+    };
+    // DeerFolia end
+
     @Override
     protected PathFinder createPathFinder(final int maxVisitedNodes) {
         this.allowBreaching = this.mob.is(EntityType.DOLPHIN);
         this.nodeEvaluator = new SwimNodeEvaluator(this.allowBreaching);
         this.nodeEvaluator.setCanPassDoors(false);
+        // DeerFolia start - async path processing
+        if (cn.lunadeer.mc.deerfolia.DeerFoliaConfiguration.asyncPathfinding.enabled) {
+            return new PathFinder(this.nodeEvaluator, maxVisitedNodes, nodeEvaluatorGenerator);
+        }
+        // DeerFolia end
         return new PathFinder(this.nodeEvaluator, maxVisitedNodes);
     }
 
diff --git a/net/minecraft/world/entity/ai/sensing/NearestBedSensor.java b/net/minecraft/world/entity/ai/sensing/NearestBedSensor.java
index f47e879913fc23e1c741366fc8e77ca15192d832..0b243158a4a940716829e31f5ff1b58968a24b9b 100644
--- a/net/minecraft/world/entity/ai/sensing/NearestBedSensor.java
+++ b/net/minecraft/world/entity/ai/sensing/NearestBedSensor.java
@@ -57,6 +57,26 @@ public class NearestBedSensor extends Sensor<Mob> {
                     e -> e.is(PoiTypes.HOME), cacheTest, body.blockPosition(), AcquirePoi.SCAN_RANGE, PoiManager.Occupancy.ANY
                 )
                 .collect(Collectors.toSet());
+            // DeerFolia start - await on async path processing
+            if (cn.lunadeer.mc.deerfolia.DeerFoliaConfiguration.asyncPathfinding.enabled) {
+                Path possiblePath = AcquirePoi.findPathToPois(body, pois);
+                cn.lunadeer.mc.deerfolia.async.path.AsyncPathProcessor.awaitProcessing(body, possiblePath, path -> {
+                    if ((path == null || !path.canReach()) && this.triedCount < 5) {
+                        this.batchCache.long2LongEntrySet().removeIf(entry -> entry.getLongValue() < this.lastUpdate);
+                        return;
+                    }
+                    if (path == null) {
+                        return;
+                    }
+
+                    BlockPos targetPos = path.getTarget();
+                    Optional<Holder<PoiType>> type = poiManager.getTypeWithoutLoad(targetPos);
+                    if (type.isPresent()) {
+                        body.getBrain().setMemory(MemoryModuleType.NEAREST_BED, targetPos);
+                    }
+                });
+            } else {
+            // DeerFolia end
             Path path = AcquirePoi.findPathToPois(body, pois);
             if (path != null && path.canReach()) {
                 BlockPos targetPos = path.getTarget();
@@ -67,6 +87,7 @@ public class NearestBedSensor extends Sensor<Mob> {
             } else if (this.triedCount < 5) {
                 this.batchCache.long2LongEntrySet().removeIf(entry -> entry.getLongValue() < this.lastUpdate);
             }
+            } // DeerFolia - async path processing
         }
     }
 }
diff --git a/net/minecraft/world/entity/ai/village/poi/PoiManager.java b/net/minecraft/world/entity/ai/village/poi/PoiManager.java
index 1d47a897c912a2aae5938beade092829bff18993..ee13d201627731bb23b35c4de322ba1343a9cf99 100644
--- a/net/minecraft/world/entity/ai/village/poi/PoiManager.java
+++ b/net/minecraft/world/entity/ai/village/poi/PoiManager.java
@@ -46,6 +46,11 @@ public class PoiManager extends SectionStorage<PoiSection, PoiSection.Packed> im
     public static final int VILLAGE_SECTION_SIZE = 1;
     private final PoiManager.DistanceTracker distanceTracker;
     private final LongSet loadedChunks = new LongOpenHashSet();
+    // DeerFolia - Async path processing
+    private final java.util.concurrent.locks.ReadWriteLock rwlock = new java.util.concurrent.locks.ReentrantReadWriteLock();
+    private final java.util.concurrent.locks.Lock rlock = this.rwlock.readLock();
+    private final java.util.concurrent.locks.Lock wlock = this.rwlock.writeLock();
+    // DeerFolia - Async path processing
 
     // Paper start - rewrite chunk system
     private final net.minecraft.server.level.ServerLevel world;
@@ -76,11 +81,24 @@ public class PoiManager extends SectionStorage<PoiSection, PoiSection.Packed> im
         final int chunkY = ca.spottedleaf.moonrise.common.util.CoordinateUtils.getChunkSectionY(pos);
         final int chunkZ = ca.spottedleaf.moonrise.common.util.CoordinateUtils.getChunkSectionZ(pos);
 
-        ca.spottedleaf.moonrise.common.util.TickThread.ensureTickThread(this.world, chunkX, chunkZ, "Accessing poi chunk off-main");
-
-        final ca.spottedleaf.moonrise.patches.chunk_system.level.poi.PoiChunk ret = ((ca.spottedleaf.moonrise.patches.chunk_system.level.ChunkSystemServerLevel)this.world).moonrise$getChunkTaskScheduler().chunkHolderManager.getPoiChunkIfLoaded(chunkX, chunkZ, true);
-
-        return ret == null ? Optional.empty() : ret.getSectionForVanilla(chunkY);
+        // DeerFolia - Async path processing
+        try {
+            if (this.rlock.tryLock(100, java.util.concurrent.TimeUnit.MILLISECONDS)) {
+                try {
+                    final ca.spottedleaf.moonrise.patches.chunk_system.level.poi.PoiChunk ret = ((ca.spottedleaf.moonrise.patches.chunk_system.level.ChunkSystemServerLevel)this.world)
+                        .moonrise$getChunkTaskScheduler()
+                        .chunkHolderManager
+                        .getPoiChunkIfLoaded(chunkX, chunkZ, true);
+                    return ret == null ? Optional.empty() : ret.getSectionForVanilla(chunkY);
+                } finally {
+                    this.rlock.unlock();
+                }
+            }
+            return Optional.empty();
+        } catch (InterruptedException e) {
+            return Optional.empty();
+        }
+        // DeerFolia - Async path processing
     }
 
     @Override
@@ -365,16 +383,40 @@ public class PoiManager extends SectionStorage<PoiSection, PoiSection.Packed> im
         final Predicate<Holder<PoiType>> predicate, final BiPredicate<Holder<PoiType>, BlockPos> filter, final BlockPos center, final int radius
     ) {
         // Paper start - optimise POI lookup
-        final PoiRecord record = ca.spottedleaf.moonrise.patches.poi_lookup.PoiAccess.findAnyPoiRecord(
-            (PoiManager)(Object)this, predicate, filter, center, radius, (double)((long)radius * (long)radius), PoiManager.Occupancy.HAS_SPACE, ca.spottedleaf.moonrise.patches.poi_lookup.PoiAccess.LOAD_FOR_SEARCHING
-        );
-
-        if (record == null) {
-            return Optional.empty();
+        // DeerFolia - Async path processing
+        try {
+            PoiRecord record = null;
+            if (this.rlock.tryLock(100, java.util.concurrent.TimeUnit.MILLISECONDS)) {
+                try {
+                    record = ca.spottedleaf.moonrise.patches.poi_lookup.PoiAccess.findAnyPoiRecord(
+                        (PoiManager)(Object)this,
+                        predicate,
+                        filter,
+                        center,
+                        radius,
+                        (double)((long)radius * (long)radius),
+                        PoiManager.Occupancy.HAS_SPACE,
+                        ca.spottedleaf.moonrise.patches.poi_lookup.PoiAccess.LOAD_FOR_SEARCHING
+                    );
+                } finally {
+                    this.rlock.unlock();
+                }
+            }
+            if (record == null) {
+                return Optional.empty();
+            }
+            if (this.wlock.tryLock(100, java.util.concurrent.TimeUnit.MILLISECONDS)) {
+                try {
+                    record.acquireTicket();
+                    return Optional.of(record.getPos());
+                } finally {
+                    this.wlock.unlock();
+                }
+            }
+        } catch (InterruptedException ignored) {
         }
-
-        record.acquireTicket();
-        return Optional.of(record.getPos());
+        return Optional.empty();
+        // DeerFolia - Async path processing
         // Paper end - optimise POI lookup
     }
 
@@ -413,6 +455,12 @@ public class PoiManager extends SectionStorage<PoiSection, PoiSection.Packed> im
         return this.getOrLoad(SectionPos.asLong(pos)).map(s -> s.exists(pos, predicate)).orElse(false);
     }
 
+    // DeerFolia - Async path processing
+    public Optional<Holder<PoiType>> getTypeWithoutLoad(final BlockPos pos) {
+        return this.get(SectionPos.asLong(pos)).flatMap(section -> section.getType(pos));
+    }
+    // DeerFolia - Async path processing
+
     public Optional<Holder<PoiType>> getType(final BlockPos pos) {
         return this.getOrLoad(SectionPos.asLong(pos)).flatMap(section -> section.getType(pos));
     }
diff --git a/net/minecraft/world/entity/animal/frog/Frog.java b/net/minecraft/world/entity/animal/frog/Frog.java
index b1612cf1ec4f0c9ec1bc39c4e5e733d5c2d5070e..2a8de8f9bb2e8bc6b09b16aaffbcbcc187a16075 100644
--- a/net/minecraft/world/entity/animal/frog/Frog.java
+++ b/net/minecraft/world/entity/animal/frog/Frog.java
@@ -390,6 +390,17 @@ public class Frog extends Animal {
             super(mob, level);
         }
 
+        // DeerFolia start - petal - async path processing
+        private static final cn.lunadeer.mc.deerfolia.async.path.NodeEvaluatorGenerator nodeEvaluatorGenerator = nodeEvaluatorFeatures -> {
+            Frog.FrogNodeEvaluator nodeEvaluator = new Frog.FrogNodeEvaluator(true);
+            nodeEvaluator.setCanPassDoors(nodeEvaluatorFeatures.canPassDoors());
+            nodeEvaluator.setCanFloat(nodeEvaluatorFeatures.canFloat());
+            nodeEvaluator.setCanWalkOverFences(nodeEvaluatorFeatures.canWalkOverFences());
+            nodeEvaluator.setCanOpenDoors(nodeEvaluatorFeatures.canOpenDoors());
+            return nodeEvaluator;
+        };
+        // DeerFolia end
+
         @Override
         public boolean canCutCorner(final PathType pathType) {
             return pathType != PathType.WATER_BORDER && super.canCutCorner(pathType);
@@ -398,6 +409,11 @@ public class Frog extends Animal {
         @Override
         protected PathFinder createPathFinder(final int maxVisitedNodes) {
             this.nodeEvaluator = new Frog.FrogNodeEvaluator(true);
+            // DeerFolia start - petal - async path processing
+            if (cn.lunadeer.mc.deerfolia.DeerFoliaConfiguration.asyncPathfinding.enabled) {
+                return new PathFinder(this.nodeEvaluator, maxVisitedNodes, nodeEvaluatorGenerator);
+            }
+            // DeerFolia end
             return new PathFinder(this.nodeEvaluator, maxVisitedNodes);
         }
     }
diff --git a/net/minecraft/world/entity/monster/Strider.java b/net/minecraft/world/entity/monster/Strider.java
index 8f58f2f9dae3809c72cf59c0da3523c8923177c2..595d47df206c22305bb923c492f1373c9ef33d67 100644
--- a/net/minecraft/world/entity/monster/Strider.java
+++ b/net/minecraft/world/entity/monster/Strider.java
@@ -516,9 +516,25 @@ public class Strider extends Animal implements ItemSteerable {
             super(mob, level);
         }
 
+        // DeerFolia start - petal - async path processing
+        private static final cn.lunadeer.mc.deerfolia.async.path.NodeEvaluatorGenerator nodeEvaluatorGenerator = nodeEvaluatorFeatures -> {
+            WalkNodeEvaluator nodeEvaluator = new WalkNodeEvaluator();
+            nodeEvaluator.setCanPassDoors(nodeEvaluatorFeatures.canPassDoors());
+            nodeEvaluator.setCanFloat(nodeEvaluatorFeatures.canFloat());
+            nodeEvaluator.setCanWalkOverFences(nodeEvaluatorFeatures.canWalkOverFences());
+            nodeEvaluator.setCanOpenDoors(nodeEvaluatorFeatures.canOpenDoors());
+            return nodeEvaluator;
+        };
+        // DeerFolia end
+
         @Override
         protected PathFinder createPathFinder(final int maxVisitedNodes) {
             this.nodeEvaluator = new WalkNodeEvaluator();
+            // DeerFolia start - async path processing
+            if (cn.lunadeer.mc.deerfolia.DeerFoliaConfiguration.asyncPathfinding.enabled) {
+                return new PathFinder(this.nodeEvaluator, maxVisitedNodes, nodeEvaluatorGenerator);
+            }
+            // DeerFolia end
             return new PathFinder(this.nodeEvaluator, maxVisitedNodes);
         }
 
diff --git a/net/minecraft/world/entity/monster/warden/Warden.java b/net/minecraft/world/entity/monster/warden/Warden.java
index bbc620d0781824c83561065882341daf6b65d406..e99648d09f6cf81d1b9fc1f92ffb51f01860eeab 100644
--- a/net/minecraft/world/entity/monster/warden/Warden.java
+++ b/net/minecraft/world/entity/monster/warden/Warden.java
@@ -560,6 +560,20 @@ public class Warden extends Monster implements VibrationSystem {
             @Override
             protected PathFinder createPathFinder(final int maxVisitedNodes) {
                 this.nodeEvaluator = new WalkNodeEvaluator();
+                // DeerFolia start - petal - async path processing
+                if (cn.lunadeer.mc.deerfolia.DeerFoliaConfiguration.asyncPathfinding.enabled) {
+                    return new PathFinder(this.nodeEvaluator, maxVisitedNodes, GroundPathNavigation.nodeEvaluatorGenerator) {
+                        {
+                            Objects.requireNonNull(nodeEvaluator);
+                        }
+
+                        @Override
+                        protected float distance(final Node from, final Node to) {
+                            return from.distanceToXZ(to);
+                        }
+                    };
+                }
+                // DeerFolia end
                 return new PathFinder(this.nodeEvaluator, maxVisitedNodes) {
                     {
                         Objects.requireNonNull(nodeEvaluator);
diff --git a/net/minecraft/world/level/chunk/LevelChunk.java b/net/minecraft/world/level/chunk/LevelChunk.java
index 5211ee4c8101d96caaff95c8a7b2932eed5567ee..ccd513b7bbe7cdf70bbbf941172f26c9fef9d64b 100644
--- a/net/minecraft/world/level/chunk/LevelChunk.java
+++ b/net/minecraft/world/level/chunk/LevelChunk.java
@@ -473,7 +473,10 @@ public class LevelChunk extends ChunkAccess implements DebugValueSource, ca.spot
 
     public @Nullable BlockEntity getBlockEntity(final BlockPos pos, final LevelChunk.EntityCreationType creationType) {
         // CraftBukkit start
-        BlockEntity blockEntity = this.level.getCurrentWorldData().capturedTileEntities.get(pos); // Folia - region threading
+        // BlockEntity blockEntity = this.level.getCurrentWorldData().capturedTileEntities.get(pos); // Folia - region threading
+        BlockEntity blockEntity = java.util.Optional.ofNullable(this.level.getCurrentWorldData())
+            .map(data -> data.capturedTileEntities.get(pos))
+            .orElse(null);
         if (blockEntity == null) {
             blockEntity = this.blockEntities.get(pos);
         }
diff --git a/net/minecraft/world/level/pathfinder/Path.java b/net/minecraft/world/level/pathfinder/Path.java
index ce536d07512c3cb9a8f56a1c4f6506db50161f8b..4aaaa67d9b64f1795224bd6be9ccf072a4ca1ce5 100644
--- a/net/minecraft/world/level/pathfinder/Path.java
+++ b/net/minecraft/world/level/pathfinder/Path.java
@@ -11,7 +11,7 @@ import net.minecraft.world.entity.Entity;
 import net.minecraft.world.phys.Vec3;
 import org.jspecify.annotations.Nullable;
 
-public final class Path {
+public class Path {
     public static final StreamCodec<FriendlyByteBuf, Path> STREAM_CODEC = StreamCodec.of((output, value) -> value.writeToStream(output), Path::createFromStream);
     public final List<Node> nodes;
     private Path.@Nullable DebugData debugData;
@@ -27,6 +27,12 @@ public final class Path {
         this.reached = reached;
     }
 
+    // DeerFolia start - petal - async path processing
+    public boolean isProcessed() {
+        return true;
+    }
+    // DeerFolia end
+
     public void advance() {
         this.nextNodeIndex++;
     }
@@ -98,6 +104,7 @@ public final class Path {
     }
 
     public boolean sameAs(final @Nullable Path path) {
+        if (path == this) return true; // DeerFolia - petal - short circuit
         return path != null && this.nodes.equals(path.nodes);
     }
 
diff --git a/net/minecraft/world/level/pathfinder/PathFinder.java b/net/minecraft/world/level/pathfinder/PathFinder.java
index 7564a67d2d63c3a7f95e1d6008e7308620653774..7e3523f902ec7b41306d0ccbcf99ab52f53cb11f 100644
--- a/net/minecraft/world/level/pathfinder/PathFinder.java
+++ b/net/minecraft/world/level/pathfinder/PathFinder.java
@@ -26,10 +26,22 @@ public class PathFinder {
     public final NodeEvaluator nodeEvaluator;
     private final BinaryHeap openSet = new BinaryHeap();
     private BooleanSupplier captureDebug = () -> false;
+    private final cn.lunadeer.mc.deerfolia.async.path.NodeEvaluatorGenerator nodeEvaluatorGenerator; // DeerFolia - petal - we use this later to generate an evaluator
 
-    public PathFinder(final NodeEvaluator nodeEvaluator, final int maxVisitedNodes) {
+    public PathFinder(
+        final NodeEvaluator nodeEvaluator,
+        final int maxVisitedNodes,
+        final cn.lunadeer.mc.deerfolia.async.path.NodeEvaluatorGenerator nodeEvaluatorGenerator
+    ) { // DeerFolia - petal - add nodeEvaluatorGenerator
         this.nodeEvaluator = nodeEvaluator;
         this.maxVisitedNodes = maxVisitedNodes;
+        // DeerFolia start - petal - support nodeEvaluatorgenerators
+        this.nodeEvaluatorGenerator = nodeEvaluatorGenerator;
+    }
+
+    public PathFinder(final NodeEvaluator nodeEvaluator, final int maxVisitedNodes) {
+        this(nodeEvaluator, maxVisitedNodes, null);
+        // DeerFolia end
     }
 
     public void setCaptureDebug(final BooleanSupplier captureDebug) {
@@ -48,27 +60,69 @@ public class PathFinder {
         final int reachRange,
         final float maxVisitedNodesMultiplier
     ) {
-        this.openSet.clear();
-        this.nodeEvaluator.prepare(level, entity);
-        Node from = this.nodeEvaluator.getStart();
+        if (!cn.lunadeer.mc.deerfolia.DeerFoliaConfiguration.asyncPathfinding.enabled)
+            this.openSet.clear(); // DeerFolia - petal - it's always cleared in processPath
+        // DeerFolia start - petal - use a generated evaluator if we have one otherwise run sync
+        NodeEvaluator nodeEvaluator = this.nodeEvaluatorGenerator == null
+            ? this.nodeEvaluator
+            : cn.lunadeer.mc.deerfolia.async.path.NodeEvaluatorCache.takeNodeEvaluator(this.nodeEvaluatorGenerator, this.nodeEvaluator);
+        nodeEvaluator.prepare(level, entity);
+        Node from = nodeEvaluator.getStart();
+        // DeerFolia end
         if (from == null) {
+            cn.lunadeer.mc.deerfolia.async.path.NodeEvaluatorCache.removeNodeEvaluator(nodeEvaluator); // DeerFolia - petal - handle nodeEvaluatorGenerator
             return null;
         } else {
             // Paper start - Perf: remove streams and optimize collection
             List<Map.Entry<Target, BlockPos>> tos = Lists.newArrayList();
             for (BlockPos pos : targets) {
-                tos.add(new java.util.AbstractMap.SimpleEntry<>(this.nodeEvaluator.getTarget(pos.getX(), pos.getY(), pos.getZ()), pos));
+                tos.add(new java.util.AbstractMap.SimpleEntry<>(nodeEvaluator.getTarget(pos.getX(), pos.getY(), pos.getZ()), pos)); // DeerFolia - petal - handle nodeEvaluatorGenerator
             }
             // Paper end - Perf: remove streams and optimize collection
-            Path path = this.findPath(from, tos, maxPathLength, reachRange, maxVisitedNodesMultiplier);
-            this.nodeEvaluator.done();
-            return path;
+            // DeerFolia start - petal - async path processing
+            if (this.nodeEvaluatorGenerator == null) {
+                cn.lunadeer.mc.deerfolia.async.path.NodeEvaluatorCache.removeNodeEvaluator(nodeEvaluator);
+                return this.findPath(from, tos, maxPathLength, reachRange, maxVisitedNodesMultiplier); // Gale - Purpur - remove vanilla profiler
+            }
+            return new cn.lunadeer.mc.deerfolia.async.path.AsyncPath(Lists.newArrayList(), targets, () -> {
+                try {
+                    return this.processPath(nodeEvaluator, from, tos, maxPathLength, reachRange, maxVisitedNodesMultiplier);
+                } catch (Exception e) {
+                    e.printStackTrace();
+                    return null;
+                } finally {
+                    nodeEvaluator.done();
+                    cn.lunadeer.mc.deerfolia.async.path.NodeEvaluatorCache.returnNodeEvaluator(nodeEvaluator);
+                }
+            });
+            // DeerFolia end
         }
     }
 
     private @Nullable Path findPath(
         final Node from, final List<Map.Entry<Target, BlockPos>> targets, final float maxPathLength, final int reachRange, final float maxVisitedNodesMultiplier // Paper - optimize collection
     ) {
+        // DeerFolia start - petal - split pathfinding into the original sync method for compat and processing for delaying
+        try {
+            return this.processPath(this.nodeEvaluator, from, targets, maxPathLength, reachRange, maxVisitedNodesMultiplier);
+        } catch (Exception e) {
+            e.printStackTrace();
+            return null;
+        } finally {
+            this.nodeEvaluator.done();
+        }
+    }
+
+    private synchronized @org.jetbrains.annotations.NotNull Path processPath(
+        final NodeEvaluator nodeEvaluator,
+        final Node from,
+        final List<Map.Entry<Target, BlockPos>> targets,
+        final float maxPathLength,
+        final int reachRange,
+        final float maxVisitedNodesMultiplier
+    ) { // sync to only use the caching functions in this class on a single thread
+        org.apache.commons.lang3.Validate.isTrue(!targets.isEmpty()); // ensure that we have at least one position, which means we'll always return a path
+        // DeerFolia end
         ProfilerFiller profiler = Profiler.get();
         profiler.push("find_path");
         profiler.markForCharting(MetricCategory.PATH_FINDING);
@@ -112,7 +166,7 @@ public class PathFinder {
             }
 
             if (!(current.distanceTo(from) >= maxPathLength)) {
-                int neighborCount = this.nodeEvaluator.getNeighbors(this.neighbors, current);
+                int neighborCount = nodeEvaluator.getNeighbors(this.neighbors, current); // DeerFolia - petal - use provided nodeEvaluator
 
                 for (int i = 0; i < neighborCount; i++) {
                     Node neighbor = this.neighbors[i];
diff --git a/net/minecraft/world/level/pathfinder/SwimNodeEvaluator.java b/net/minecraft/world/level/pathfinder/SwimNodeEvaluator.java
index c61bcf31413342259a31d3bc1451ea6968b1e46b..5e50e4a6af85eeabc8b05775f9be56d99265bbf5 100644
--- a/net/minecraft/world/level/pathfinder/SwimNodeEvaluator.java
+++ b/net/minecraft/world/level/pathfinder/SwimNodeEvaluator.java
@@ -15,7 +15,7 @@ import net.minecraft.world.level.material.FluidState;
 import org.jspecify.annotations.Nullable;
 
 public class SwimNodeEvaluator extends NodeEvaluator {
-    private final boolean allowBreaching;
+    public final boolean allowBreaching; // DeerFolia - make this public
     private final Long2ObjectMap<PathType> pathTypesByPosCache = new Long2ObjectOpenHashMap<>();
 
     public SwimNodeEvaluator(final boolean allowBreaching) {
