From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: zhangyuheng <zhangyuheng@lunadeer.cn>
Date: Fri, 27 Feb 2026 11:54:41 +0800
Subject: [PATCH] Fix Mob getTarget race NPE


diff --git a/net/minecraft/world/entity/Mob.java b/net/minecraft/world/entity/Mob.java
index c8a9f0ff3185b98342f573a8ae9eec7eae63a4e9..e337eafc819e2aa57cf35a9107dd6664407c183b 100644
--- a/net/minecraft/world/entity/Mob.java
+++ b/net/minecraft/world/entity/Mob.java
@@ -260,10 +260,13 @@ public abstract class Mob extends LivingEntity implements Targeting, EquipmentUs
     @Override
     public @Nullable LivingEntity getTarget() {
         // Folia start - region threading
-        if (this.target != null && (!ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(this.target) || this.target.isRemoved())) {
+        // DeerFolia start - avoid race NPE on target
+        LivingEntity target = this.target;
+        if (target != null && (!ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(target) || target.isRemoved())) {
             this.target = null;
             return null;
         }
+        // DeerFolia end - avoid race NPE on target
         // Folia end - region threading
         return this.asValidTarget(this.target);
     }
