From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: zhangyuheng <zhangyuheng@lunadeer.cn>
Date: Fri, 27 Feb 2026 11:54:47 +0800
Subject: [PATCH] Fix Path sameAs ConcurrentModificationException with async
 pathfinding


diff --git a/net/minecraft/world/level/pathfinder/Path.java b/net/minecraft/world/level/pathfinder/Path.java
index 4aaaa67d9b64f1795224bd6be9ccf072a4ca1ce5..46c5a94a7912b31ecbe22427534f89c794942db4 100644
--- a/net/minecraft/world/level/pathfinder/Path.java
+++ b/net/minecraft/world/level/pathfinder/Path.java
@@ -105,6 +105,7 @@ public class Path {
 
     public boolean sameAs(final @Nullable Path path) {
         if (path == this) return true; // DeerFolia - petal - short circuit
+        if (!this.isProcessed() || (path != null && !path.isProcessed())) return false; // DeerFolia - avoid ConcurrentModificationException with async pathfinding
         return path != null && this.nodes.equals(path.nodes);
     }
 
